// baqend/commonjs/lib/util/Lockable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lockable = void 0;
/**
 * This base class provides an lock interface to execute exclusive operations
 */
var Lockable = /** @class */ (function () {
    function Lockable() {
        /**
         * Indicates if there is currently an onging exclusive operation
         * @type boolean
         * @private
         */
        this.isLocked = false;
        /**
         * A promise which represents the state of the least exclusive operation
         * @type Promise
         * @private
         */
        this.readyPromise = Promise.resolve(this);
    }
    Object.defineProperty(Lockable.prototype, "isReady", {
        /**
         * Indicates if there is currently no exclusive operation executed
         * <code>true</code> If no exclusive lock is hold
         */
        get: function () {
            return !this.isLocked;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Waits on the previously requested operation and calls the doneCallback if the operation is fulfilled
     * @param doneCallback The callback which will be invoked when the previously
     * operations on this object is completed.
     * @param failCallback When the lock can't be released caused by a none
     * recoverable error
     * @return A promise which completes successfully, when the previously requested
     * operation completes
     */
    Lockable.prototype.ready = function (doneCallback, failCallback) {
        return this.readyPromise.then(doneCallback, failCallback);
    };
    /**
     * Try to aquire an exclusive lock and executes the given callback.
     * @param callback The exclusive operation to execute
     * @param [critical=false] Indicates if the operation is critical. If the operation is critical and the
     * operation fails, then the lock will not be released
     * @return A promise
     * @throws If the lock can't be aquired
     * @protected
     */
    Lockable.prototype.withLock = function (callback, critical) {
        var _this = this;
        if (critical === void 0) { critical = false; }
        if (this.isLocked) {
            throw new Error('Current operation has not been finished.');
        }
        try {
            this.isLocked = true;
            var result = callback().then(function (res) {
                _this.isLocked = false;
                return res;
            }, function (e) {
                if (!critical) {
                    _this.isLocked = false;
                }
                throw e;
            });
            this.readyPromise = result.then(function () { return _this; }, function (e) {
                if (!critical) {
                    return _this;
                }
                throw e;
            });
            return result;
        }
        catch (e) {
            if (critical) {
                this.readyPromise = Promise.reject(e);
            }
            else {
                this.isLocked = false;
            }
            throw e;
        }
    };
    return Lockable;
}());
exports.Lockable = Lockable;
