// baqend/esm/lib/query/Builder.js
import { Filter } from './Filter';
import { Condition } from './Condition';
import { Operator } from './Operator';
import { Query, flatArgs, } from './Query';
import { Node } from './Node';
export class Builder extends Query {
    /**
     * Joins the conditions by an logical AND
     * @param args The query nodes to join
     * @return Returns a new query which joins the given queries by a logical AND
     */
    and(...args) {
        return this.addOperator('$and', flatArgs(args));
    }
    /**
     * Joins the conditions by an logical OR
     * @param args The query nodes to join
     * @return Returns a new query which joins the given queries by a logical OR
     */
    or(...args) {
        return this.addOperator('$or', flatArgs(args));
    }
    /**
     * Joins the conditions by an logical NOR
     * @param args The query nodes to join
     * @return Returns a new query which joins the given queries by a logical NOR
     */
    nor(...args) {
        return this.addOperator('$nor', flatArgs(args));
    }
    eventStream(options, onNext, onError, onComplete) {
        return this.where({}).eventStream(options, onNext, onError, onComplete);
    }
    resultStream(options, onNext, onError, onComplete) {
        return this.where({}).resultStream(options, onNext, onError, onComplete);
    }
    /**
     * @inheritDoc
     */
    resultList(options, doneCallback, failCallback) {
        return this.where({}).resultList(options, doneCallback, failCallback);
    }
    /**
     * @inheritDoc
     */
    singleResult(options, doneCallback, failCallback) {
        return this.where({}).singleResult(options, doneCallback, failCallback);
    }
    /**
     * @inheritDoc
     */
    count(doneCallback, failCallback) {
        return this.where({}).count(doneCallback, failCallback);
    }
    addOperator(operator, args) {
        if (args.length < 2) {
            throw new Error(`Only two or more queries can be joined with an ${operator} operator.`);
        }
        args.forEach((arg, index) => {
            if (!(arg instanceof Node)) {
                throw new Error(`Argument at index ${index} is not a query.`);
            }
        });
        return new Operator(this.entityManager, this.resultClass, operator, args);
    }
    addOrder(fieldOrSort, order) {
        return new Filter(this.entityManager, this.resultClass).addOrder(fieldOrSort, order);
    }
    addFilter(field, filter, value) {
        return new Filter(this.entityManager, this.resultClass).addFilter(field, filter, value);
    }
    addOffset(offset) {
        return new Filter(this.entityManager, this.resultClass).addOffset(offset);
    }
    addLimit(limit) {
        return new Filter(this.entityManager, this.resultClass).addLimit(limit);
    }
}
Object.assign(Builder.prototype, Condition);
