// baqend/esm/lib/query/Condition.js
import { flatArgs } from './Query';
export const Condition = {
    where(conditions) {
        return this.addFilter(null, null, conditions);
    },
    equal(field, value) {
        return this.addFilter(field, null, value);
    },
    notEqual(field, value) {
        return this.addFilter(field, '$ne', value);
    },
    greaterThan(field, value) {
        return this.addFilter(field, '$gt', value);
    },
    greaterThanOrEqualTo(field, value) {
        return this.addFilter(field, '$gte', value);
    },
    lessThan(field, value) {
        return this.addFilter(field, '$lt', value);
    },
    lessThanOrEqualTo(field, value) {
        return this.addFilter(field, '$lte', value);
    },
    between(field, greaterValue, lessValue) {
        return this
            .addFilter(field, '$gt', greaterValue)
            .addFilter(field, '$lt', lessValue);
    },
    in(field, ...args) {
        return this.addFilter(field, '$in', flatArgs(args));
    },
    notIn(field, ...args) {
        return this.addFilter(field, '$nin', flatArgs(args));
    },
    isNull(field) {
        return this.equal(field, null);
    },
    isNotNull(field) {
        return this.addFilter(field, '$exists', true)
            .addFilter(field, '$ne', null);
    },
    containsAll(field, ...args) {
        return this.addFilter(field, '$all', flatArgs(args));
    },
    mod(field, divisor, remainder) {
        return this.addFilter(field, '$mod', [divisor, remainder]);
    },
    matches(field, regExp) {
        const reg = regExp instanceof RegExp ? regExp : new RegExp(regExp);
        if (reg.ignoreCase) {
            throw new Error('RegExp.ignoreCase flag is not supported.');
        }
        if (reg.global) {
            throw new Error('RegExp.global flag is not supported.');
        }
        if (reg.source.indexOf('^') !== 0) {
            throw new Error('regExp must be an anchored expression, i.e. it must be started with a ^.');
        }
        const result = this.addFilter(field, '$regex', reg.source);
        if (reg.multiline) {
            result.addFilter(field, '$options', 'm');
        }
        return result;
    },
    size(field, size) {
        return this.addFilter(field, '$size', size);
    },
    near(field, geoPoint, maxDistance) {
        return this.addFilter(field, '$nearSphere', {
            $geometry: {
                type: 'Point',
                coordinates: [geoPoint.longitude, geoPoint.latitude],
            },
            $maxDistance: maxDistance,
        });
    },
    withinPolygon(field, ...args) {
        const geoPoints = flatArgs(args);
        return this.addFilter(field, '$geoWithin', {
            $geometry: {
                type: 'Polygon',
                coordinates: [geoPoints.map((geoPoint) => [geoPoint.longitude, geoPoint.latitude])],
            },
        });
    },
};
// aliases
Object.assign(Condition, {
    eq: Condition.equal,
    ne: Condition.notEqual,
    lt: Condition.lessThan,
    le: Condition.lessThanOrEqualTo,
    gt: Condition.greaterThan,
    ge: Condition.greaterThanOrEqualTo,
    containsAny: Condition.in,
});
