// baqend/esm/lib/connector/WebSocketConnector.js
import { CommunicationError } from '../error';
import { WebSocket } from '../util/websocket';
import { Observable } from '../util/observable';
export class WebSocketConnector {
    /**
     *url
     */
    constructor(url) {
        this.observers = {};
        this.socket = null;
        this.url = url;
    }
    /**
     *url The websocket connect script url
     *a websocket connection
     */
    static create(url) {
        let websocket = this.websockets[url];
        if (!websocket) {
            websocket = new WebSocketConnector(url);
            this.websockets[url] = websocket;
        }
        return websocket;
    }
    open() {
        if (!this.socket) {
            const socket = new WebSocket(this.url);
            let socketPromise;
            const handleSocketCompletion = (error) => {
                // observable error calls can throw an exception therefore cleanup beforehand
                let isError = false;
                if (this.socket === socketPromise) {
                    isError = socket.readyState !== 3;
                    this.socket = null;
                }
                let firstErr = null;
                Object.keys(this.observers).forEach((id) => {
                    const observer = this.observers[id];
                    delete this.observers[id]; // unsubscribe to allow re subscriptions
                    if (!observer) {
                        return;
                    }
                    try {
                        if (isError) {
                            observer.error(new CommunicationError(null, Object.assign({ status: 0, headers: {} }, (error instanceof Error && { error }))));
                        }
                        else {
                            observer.complete();
                        }
                    }
                    catch (e) {
                        if (!firstErr) {
                            firstErr = e;
                        }
                    }
                });
                if (firstErr) {
                    throw firstErr;
                }
            };
            socket.onerror = handleSocketCompletion;
            socket.onclose = handleSocketCompletion;
            socket.onmessage = (event) => {
                const message = JSON.parse(event.data);
                message.date = new Date(message.date);
                const { id } = message;
                if (!id) {
                    if (message.type === 'error') {
                        handleSocketCompletion(message);
                    }
                    return;
                }
                const observer = this.observers[id];
                if (observer) {
                    if (message.type === 'error') {
                        observer.error(new CommunicationError(null, message));
                    }
                    else {
                        observer.next(message);
                    }
                }
            };
            socketPromise = new Promise((resolve) => {
                socket.onopen = resolve;
            }).then(() => socket);
            this.socket = socketPromise;
        }
        return this.socket;
    }
    close() {
        if (this.socket) {
            this.socket.then((socket) => {
                socket.close();
            });
            this.socket = null;
        }
    }
    /**
     *tokenStorage
     *id subscription ID
     *The channel for sending and receiving messages
     */
    openStream(tokenStorage, id) {
        const stream = new Observable((observer) => {
            if (this.observers[id]) {
                throw new Error('Only one subscription per stream is allowed.');
            }
            this.observers[id] = observer;
            return () => {
                // cleanup only our subscription and handle re subscription on the same stream id correctly
                if (this.observers[id] === observer) {
                    delete this.observers[id];
                }
            };
        });
        Object.assign(stream, {
            send: (message) => {
                this.open().then((socket) => {
                    const jsonMessage = JSON.stringify(Object.assign(Object.assign({ id }, message), (tokenStorage.token && { token: tokenStorage.token })));
                    socket.send(jsonMessage);
                });
            },
        });
        return stream;
    }
}
/**
 * Map of all available connectors to their respective websocket connections
 */
WebSocketConnector.websockets = {};
