// baqend/commonjs/lib/Acl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Acl = void 0;
var Permission_1 = require("./intersection/Permission");
/**
 * Creates a new Acl object, with an empty rule set for an object
 */
var Acl = /** @class */ (function () {
    function Acl() {
        /**
         * The read permission of the object
         */
        this.read = new Permission_1.Permission();
        /**
         * The write permission of the object
         */
        this.write = new Permission_1.Permission();
    }
    /**
     * Removes all acl rules, read and write access is public afterwards
     *
     * @return
     * */
    Acl.prototype.clear = function () {
        this.read.clear();
        this.write.clear();
    };
    /**
     * Copies permissions from another ACL
     *
     * @param acl The ACL to copy from
     * @return
     */
    Acl.prototype.copy = function (acl) {
        this.read.copy(acl.read);
        this.write.copy(acl.write);
        return this;
    };
    /**
     * Gets whenever all users and roles have the permission to read the object
     *
     * @return <code>true</code> If public access is allowed
     */
    Acl.prototype.isPublicReadAllowed = function () {
        return this.read.isPublicAllowed();
    };
    /**
     * Sets whenever all users and roles should have the permission to read the object
     *
     * Note: All other allow read rules will be removed.
     *
     * @return
     * */
    Acl.prototype.setPublicReadAllowed = function () {
        return this.read.setPublicAllowed();
    };
    /**
     * Checks whenever the user or role is explicit allowed to read the object
     *
     * @param userOrRole The user or role to check for
     * @return <code>true</code> if read access is explicitly allowed for the given user or role
     */
    Acl.prototype.isReadAllowed = function (userOrRole) {
        return this.read.isAllowed(userOrRole);
    };
    /**
     * Checks whenever the user or role is explicit denied to read the object
     *
     * @param userOrRole The user or role to check for
     * @return <code>true</code> if read access is explicitly denied for the given user or role
     */
    Acl.prototype.isReadDenied = function (userOrRole) {
        return this.read.isDenied(userOrRole);
    };
    /**
     * Allows the given user or rule to read the object
     *
     * @param userOrRole The user or role to allow
     * @return this acl object
     */
    Acl.prototype.allowReadAccess = function () {
        var _a;
        var userOrRole = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            userOrRole[_i] = arguments[_i];
        }
        (_a = this.read).allowAccess.apply(_a, userOrRole);
        return this;
    };
    /**
     * Denies the given user or rule to read the object
     *
     * @param userOrRole The user or role to deny
     * @return this acl object
     */
    Acl.prototype.denyReadAccess = function () {
        var _a;
        var userOrRole = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            userOrRole[_i] = arguments[_i];
        }
        (_a = this.read).denyAccess.apply(_a, userOrRole);
        return this;
    };
    /**
     * Deletes any read allow/deny rule for the given user or role
     *
     * @param userOrRole The user or role
     * @return this acl object
     */
    Acl.prototype.deleteReadAccess = function () {
        var _a;
        var userOrRole = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            userOrRole[_i] = arguments[_i];
        }
        (_a = this.read).deleteAccess.apply(_a, userOrRole);
        return this;
    };
    /**
     * Gets whenever all users and roles have the permission to write the object
     *
     * @return <code>true</code> If public access is allowed
     */
    Acl.prototype.isPublicWriteAllowed = function () {
        return this.write.isPublicAllowed();
    };
    /**
     * Sets whenever all users and roles should have the permission to write the object
     *
     * Note: All other allow write rules will be removed.
     *
     * @return
     * */
    Acl.prototype.setPublicWriteAllowed = function () {
        return this.write.setPublicAllowed();
    };
    /**
     * Checks whenever the user or role is explicit allowed to write the object
     *
     * @param userOrRole The user or role to check for
     * @return <code>true</code> if write access is explicitly allowed for the given user or role
     */
    Acl.prototype.isWriteAllowed = function (userOrRole) {
        return this.write.isAllowed(userOrRole);
    };
    /**
     * Checks whenever the user or role is explicit denied to write the object
     *
     * @param userOrRole The user or role to check for
     * @return <code>true</code> if write access is explicitly denied for the given user or role
     */
    Acl.prototype.isWriteDenied = function (userOrRole) {
        return this.write.isDenied(userOrRole);
    };
    /**
     * Allows the given user or rule to write the object
     *
     * @param userOrRole The user or role to allow
     * @return this acl object
     */
    Acl.prototype.allowWriteAccess = function () {
        var _a;
        var userOrRole = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            userOrRole[_i] = arguments[_i];
        }
        (_a = this.write).allowAccess.apply(_a, userOrRole);
        return this;
    };
    /**
     * Denies the given user or rule to write the object
     *
     * @param userOrRole The user or role to deny
     * @return this acl object
     */
    Acl.prototype.denyWriteAccess = function () {
        var _a;
        var userOrRole = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            userOrRole[_i] = arguments[_i];
        }
        (_a = this.write).denyAccess.apply(_a, userOrRole);
        return this;
    };
    /**
     * Deletes any write allow/deny rule for the given user or role
     *
     * @param userOrRole The user or role
     * @return this acl object
     */
    Acl.prototype.deleteWriteAccess = function () {
        var _a;
        var userOrRole = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            userOrRole[_i] = arguments[_i];
        }
        (_a = this.write).deleteAccess.apply(_a, userOrRole);
        return this;
    };
    /**
     * A JSON representation of the set of rules
     *
     * @return
     */
    Acl.prototype.toJSON = function () {
        return {
            read: this.read.toJSON(),
            write: this.write.toJSON(),
        };
    };
    /**
     * Sets the acl rules form JSON
     *
     * @param json The json encoded acls
     * @return
     */
    Acl.prototype.fromJSON = function (json) {
        this.read.fromJSON(json.read || {});
        this.write.fromJSON(json.write || {});
    };
    return Acl;
}());
exports.Acl = Acl;
