// baqend/commonjs/lib/deperecated-exports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.connect = exports.configure = void 0;
var binding = __importStar(require("./binding"));
var connector = __importStar(require("./connector"));
var error = __importStar(require("./error"));
var message = __importStar(require("./message"));
var util = __importStar(require("./util"));
var caching = __importStar(require("./caching"));
var query = __importStar(require("./query"));
var partialupdate = __importStar(require("./partialupdate"));
var intersection = __importStar(require("./intersection"));
var metamodel = __importStar(require("./metamodel"));
var intersection_1 = require("./intersection");
var metamodel_1 = require("./metamodel");
var EntityManagerFactory_1 = require("./EntityManagerFactory");
var EntityManager_1 = require("./EntityManager");
var Acl_1 = require("./Acl");
var baqend_1 = require("./baqend");
var util_1 = require("./util");
function deprecateExports(target, targetName, newImportSignature, exports) {
    Object.keys(exports).forEach(function (exported) {
        var decorate = util_1.deprecated(newImportSignature.replace('$export', exported));
        Object.defineProperty(target, exported, decorate(targetName, exported, {
            get: function () {
                return exports[exported];
            },
        }));
    });
}
deprecateExports(util, 'util', 'intersection.$export', {
    Permission: intersection_1.Permission, Metadata: intersection_1.Metadata, TokenStorage: intersection_1.TokenStorage, Validator: intersection_1.Validator, PushMessage: intersection_1.PushMessage, Code: intersection_1.Code, Modules: intersection_1.Modules, Logger: intersection_1.Logger,
});
deprecateExports(EntityManager_1.EntityManager.prototype, 'db', 'import { $export } from \'baqend\'', {
    db: baqend_1.db,
    binding: binding,
    connector: connector,
    error: error,
    message: message,
    util: util,
    caching: caching,
    query: query,
    partialupdate: partialupdate,
    intersection: intersection,
    EntityManagerFactory: EntityManagerFactory_1.EntityManagerFactory,
    EntityManager: EntityManager_1.EntityManager,
    Acl: Acl_1.Acl,
});
deprecateExports(metamodel_1.Metamodel.prototype, 'metamodel', 'import { metamodel } from \'baqend\';', metamodel);
function configure() {
    throw new Error('Please use Baqend.db.configure() or import { db } from \'baqend\' instead.');
}
exports.configure = configure;
function connect() {
    throw new Error('Please use Baqend.db.connect() or import { db } from \'baqend\' instead.');
}
exports.connect = connect;
