// baqend/commonjs/lib/intersection/Code.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Code = void 0;
var message = __importStar(require("../message"));
var connector_1 = require("../connector");
var Validator_1 = require("./Validator");
/**
 * Representation of a Code which runs on Baqend.
 */
var Code = /** @class */ (function () {
    /**
     * @param metamodel
     * @param entityManagerFactory
     */
    function Code(metamodel, entityManagerFactory) {
        this.metamodel = metamodel;
        this.entityManagerFactory = entityManagerFactory;
    }
    /**
     * Converts the given function to a string
     * @param fn The JavaScript function to serialize
     * @return The serialized function
     */
    Code.prototype.functionToString = function (fn) {
        if (!fn) {
            return '';
        }
        var str = fn.toString();
        str = str.substring(str.indexOf('{') + 1, str.lastIndexOf('}'));
        if (str.charAt(0) === '\n') {
            str = str.substring(1);
        }
        if (str.charAt(str.length - 1) === '\n') {
            str = str.substring(0, str.length - 1);
        }
        return str;
    };
    /**
     * Converts the given string to a module wrapper function
     * @param signature The expected parameters of the function
     * @param code The JavaScript function to deserialize
     * @return The deserialized function
     */
    Code.prototype.stringToFunction = function (signature, code) {
        // eslint-disable-next-line @typescript-eslint/no-implied-eval
        return new Function(signature /* typings are incorrect here */, code); // eslint-disable-line no-new-func
    };
    /**
     * Loads a list of all available modules without handlers
     *
     * @return
     */
    Code.prototype.loadModules = function () {
        var msg = new message.GetAllModules();
        return this.entityManagerFactory.send(msg)
            .then(function (response) { return response.entity; });
    };
    Code.prototype.loadCode = function (type, codeType, asFunction) {
        var _this = this;
        if (asFunction === void 0) { asFunction = false; }
        var bucket = typeof type === 'string' ? type : type.name;
        var msg = new message.GetBaqendCode(bucket, codeType)
            .responseType('text');
        return this.entityManagerFactory.send(msg)
            .then(function (response) { return _this.parseCode(bucket, codeType, asFunction, response.entity); }, function (e) {
            if (e.status === connector_1.StatusCode.OBJECT_NOT_FOUND) {
                return null;
            }
            throw e;
        });
    };
    Code.prototype.saveCode = function (type, codeType, fn) {
        var _this = this;
        var bucket = typeof type === 'string' ? type : type.name;
        var msg = new message.SetBaqendCode(bucket, codeType)
            .entity(fn instanceof Function ? this.functionToString(fn) : fn, 'text')
            .responseType('text');
        return this.entityManagerFactory.send(msg)
            .then(function (response) { return _this.parseCode(bucket, codeType, fn instanceof Function, response.entity); });
    };
    /**
     * Deletes Baqend code identified by the given bucket and code type
     *
     * @param type The entity type for the handler or the Name of the
     * Baqend code
     * @param codeType The type of the code
     * @return succeed if the code was deleted
     */
    Code.prototype.deleteCode = function (type, codeType) {
        var _this = this;
        var bucket = typeof type === 'string' ? type : type.name;
        var msg = new message.DeleteBaqendCode(bucket, codeType);
        return this.entityManagerFactory.send(msg)
            .then(function () { return _this.parseCode(bucket, codeType, false, null); });
    };
    /**
     * @param bucket
     * @param codeType
     * @param [asFunction=false]
     * @param code
     * @return
     * @private
     */
    Code.prototype.parseCode = function (bucket, codeType, asFunction, code) {
        if (codeType === 'validate') {
            var type = this.metamodel.entity(bucket);
            type.validationCode = code === null ? null : Validator_1.Validator.compile(type, code);
            return asFunction ? type.validationCode : code;
        }
        return code && asFunction ? this.stringToFunction(['module', 'exports'], code) : code;
    };
    return Code;
}());
exports.Code = Code;
