// baqend/commonjs/lib/intersection/PushMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PushMessage = void 0;
var binding_1 = require("../binding");
/**
 * PushMessages are used to send a push notification to a set of devices
 */
var PushMessage = /** @class */ (function () {
    /**
     * Push message will be used to send a push notification to a set of devices
     *
     * @param [devices] The Set of device references which
     * will receive this push notification.
     * @param message The message of the push notification.
     * @param subject The subject of the push notification.
     * @param [options] The options object which can contain additional information and data
     * @param [badge] The badge for iOS or Web Push devices
     * @param [data] The data object which can contain additional information.
     */
    function PushMessage(devices, message, subject, options, badge, data) {
        var opts = typeof options === 'string' ? { sound: options, badge: badge, data: data } : (options || {});
        this.devices = PushMessage.initDevices(devices);
        this.message = message;
        this.subject = subject;
        Object.assign(this, opts);
    }
    /**
     * Instantiates a set of devices from the given parameter
     * @param devices
     * @return
     */
    PushMessage.initDevices = function (devices) {
        if (devices instanceof Set) {
            return devices;
        }
        if (devices instanceof binding_1.Entity) {
            return new Set([devices]);
        }
        if (!devices || devices[Symbol.iterator]) {
            return new Set(devices);
        }
        throw new Error('Only Sets, Lists and Arrays can be used as devices.');
    };
    /**
     * Adds a new object to the set of devices
     * @param device will be added to the device set to receive the push notification
     * @return
     */
    PushMessage.prototype.addDevice = function (device) {
        this.devices.add(device);
    };
    /**
     * Converts the push message to JSON
     * @return
     */
    PushMessage.prototype.toJSON = function () {
        if (!this.devices || !this.devices.size) {
            throw new Error('Set of devices is empty.');
        }
        return Object.assign({}, this, {
            devices: Array.from(this.devices, function (device) { return device.id; }),
        });
    };
    return PushMessage;
}());
exports.PushMessage = PushMessage;
