// baqend/commonjs/lib/intersection/TokenStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenStorage = void 0;
var util_1 = require("../util");
var TokenStorage = /** @class */ (function () {
    /**
     * @param origin The origin where the token belongs to
     * @param token The initial token
     * @param temporary If the token should be saved temporary or permanently
     */
    function TokenStorage(origin, token, temporary) {
        this.tokenData = token ? TokenStorage.parse(token) : null;
        this.origin = origin;
        this.temporary = !!temporary;
    }
    /**
     * Parse a token string in its components
     * @param token The token string to parse, time values are returned as timestamps
     * @return The parsed token data
     */
    TokenStorage.parse = function (token) {
        return {
            val: token,
            createdAt: parseInt(token.substring(0, 8), 16) * 1000,
            expireAt: parseInt(token.substring(8, 16), 16) * 1000,
            sig: token.substring(token.length - 40),
            data: token.substring(0, token.length - 40),
        };
    };
    Object.defineProperty(TokenStorage.prototype, "token", {
        /**
         * Get the stored token
         * @return The token or undefined, if no token is available
         */
        get: function () {
            return this.tokenData ? this.tokenData.val : null;
        },
        enumerable: false,
        configurable: true
    });
    TokenStorage.create = function (origin) {
        return Promise.resolve(new TokenStorage(origin));
    };
    /**
     * Use the underlying storage implementation to save the token
     * @param origin The origin where the token belongs to
     * @param token The initial token
     * @param temporary If the token should be saved temporary or permanently
     * @return
     */
    TokenStorage.prototype.saveToken = function (origin, token, temporary) {
        // eslint-disable-next-line no-underscore-dangle
        if (this._saveToken !== TokenStorage.prototype._saveToken) {
            // eslint-disable-next-line no-console
            console.log('Using deprecated TokenStorage._saveToken implementation.');
            // eslint-disable-next-line no-underscore-dangle
            this._saveToken(origin, token, temporary);
        }
    };
    /**
     * Use the underlying storage implementation to save the token
     * @param origin The origin where the token belongs to
     * @param token The initial token
     * @param temporary If the token should be saved temporary or permanently
     * @return
     * @deprecated Use TokenStorage#saveToken instead
     * @protected
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    TokenStorage.prototype._saveToken = function (origin, token, temporary) { };
    /**
     * Update the token for the givin origin, the operation may be asynchronous
     * @param token The token to store or <code>null</code> to remove the token
     */
    TokenStorage.prototype.update = function (token) {
        var t = token ? TokenStorage.parse(token) : null;
        if (this.tokenData && t && this.tokenData.expireAt > t.expireAt) {
            // an older token was fetched from the cache, so ignore it
            return;
        }
        this.tokenData = t;
        this.saveToken(this.origin, token, this.temporary);
    };
    /**
     * Derives a resource token from the stored origin token and signs the resource with the generated resource token
     *
     * @param resource The resource which will be accessible with the returned token
     * @param sign Sign the given resource with a token, if sign is false the resource will only be encoded to a path
     * @return A resource token which can only be used to access the specified resource
     */
    TokenStorage.prototype.signPath = function (resource, sign) {
        if (sign === void 0) { sign = true; }
        var tokenData = this.tokenData;
        var result = Promise.resolve(resource.split('/').map(encodeURIComponent).join('/'));
        if (!tokenData || !sign) {
            return result;
        }
        return result.then(function (path) { return TokenStorage.hmac(path + tokenData.data, tokenData.sig)
            .then(function (hash) { return path + "?BAT=" + (tokenData.data + hash); }); })
            .catch(function (e) {
            // eslint-disable-next-line no-console
            console.warn('Can\'t sign the resource, run the SDK on a secured origin, or provide an alternative hmac implementation on TokenStorage.hmac', e);
            return result;
        });
    };
    TokenStorage.hmac = util_1.hmac;
    return TokenStorage;
}());
exports.TokenStorage = TokenStorage;
