// baqend/esm/lib/metamodel/BasicType.js
import { File } from '../binding';
import { GeoPoint } from '../GeoPoint';
import { PersistenceType, Type } from './Type';
function dateToJson(value) {
    // remove trailing zeros
    return value instanceof Date ? value.toISOString().replace(/\.?0*Z/, 'Z') : null;
}
function jsonToDate(json, currentValue) {
    const date = typeof json === 'string' ? new Date(json) : null;
    if (currentValue && date) {
        // compare normalized date strings instead of plain strings
        return currentValue.toISOString() === date.toISOString() ? currentValue : date;
    }
    return date;
}
export class BasicType extends Type {
    /**
     * Creates a new instance of a native db type
     * @param ref The db ref of this type
     * @param typeConstructor The javascript class of this type
     * @param noResolving Indicates if this type is not the main type of the constructor
     */
    constructor(ref, typeConstructor, noResolving) {
        const id = ref.indexOf('/db/') === 0 ? ref : `/db/${ref}`;
        super(id, typeConstructor);
        this.noResolving = !!noResolving;
    }
    /**
     * @inheritDoc
     */
    get persistenceType() {
        return PersistenceType.BASIC;
    }
    /**
     * @inheritDoc
     */
    toJsonValue(state, currentValue) {
        return currentValue === null || currentValue === undefined ? null : this.typeConstructor(currentValue);
    }
    /**
     * @inheritDoc
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    fromJsonValue(state, json, currentValue) {
        return json === null || json === undefined ? null : json;
    }
    toString() {
        return `BasicType(${this.ref})`;
    }
}
BasicType.Boolean = new class BooleanType extends BasicType {
    fromJsonValue(state, json, currentValue) {
        return typeof json === 'string' ? json !== 'false' : super.fromJsonValue(state, json, currentValue);
    }
}('Boolean', Boolean);
BasicType.Double = new class DoubleType extends BasicType {
    fromJsonValue(state, json, currentValue) {
        return typeof json === 'string' ? parseFloat(json) : super.fromJsonValue(state, json, currentValue);
    }
}('Double', Number);
BasicType.Integer = new class IntegerType extends BasicType {
    fromJsonValue(state, json, currentValue) {
        return typeof json === 'string' ? parseInt(json, 10) : super.fromJsonValue(state, json, currentValue);
    }
}('Integer', Number);
BasicType.String = new class StringType extends BasicType {
}('String', String);
BasicType.DateTime = new class DateTimeType extends BasicType {
    toJsonValue(state, value) {
        return dateToJson(value);
    }
    fromJsonValue(state, json, currentValue) {
        return jsonToDate(json, currentValue);
    }
}('DateTime', Date);
BasicType.Date = new class DateType extends BasicType {
    toJsonValue(state, value) {
        const json = dateToJson(value);
        return json ? json.substring(0, json.indexOf('T')) : null;
    }
    fromJsonValue(state, json, currentValue) {
        return jsonToDate(json, currentValue);
    }
}('Date', Date);
BasicType.Time = new class TimeType extends BasicType {
    toJsonValue(state, value) {
        const json = dateToJson(value);
        return json ? json.substring(json.indexOf('T') + 1) : null;
    }
    fromJsonValue(state, json, currentValue) {
        return typeof json === 'string' ? jsonToDate(`1970-01-01T${json}`, currentValue) : null;
    }
}('Time', Date);
BasicType.File = new class FileType extends BasicType {
    toJsonValue(state, value) {
        return value instanceof File ? value.id : null;
    }
    fromJsonValue(state, json, currentValue) {
        if (!json) {
            return null;
        }
        if (currentValue && currentValue.id === json) {
            return currentValue;
        }
        if (state.db) {
            return new state.db.File(json);
        }
        return null;
    }
}('File', File);
BasicType.GeoPoint = new class GeoPointType extends BasicType {
    toJsonValue(state, value) {
        return value instanceof GeoPoint ? value.toJSON() : null;
    }
    fromJsonValue(state, json) {
        return json ? new GeoPoint(json) : null;
    }
}('GeoPoint', GeoPoint);
BasicType.JsonArray = new class JsonArrayType extends BasicType {
    toJsonValue(state, value) {
        return Array.isArray(value) ? value : null;
    }
    fromJsonValue(state, json) {
        return Array.isArray(json) ? json : null;
    }
}('JsonArray', Array);
BasicType.JsonObject = new class JsonObjectType extends BasicType {
    toJsonValue(state, value) {
        if (value && value.constructor === Object) {
            return value;
        }
        return null;
    }
}('JsonObject', Object);
