// baqend/esm/lib/metamodel/ListAttribute.js
import { CollectionType, PluralAttribute } from './PluralAttribute';
import { Attribute } from './Attribute';
export class ListAttribute extends PluralAttribute {
    /**
     * Get the type id for this list type
     */
    static get ref() {
        return '/db/collection.List';
    }
    /**
     * @inheritDoc
     */
    get collectionType() {
        return CollectionType.LIST;
    }
    /**
     * @param name
     * @param elementType
     */
    constructor(name, elementType) {
        super(name, Array, elementType);
    }
    /**
     * @inheritDoc
     */
    getJsonValue(state, object, options) {
        const value = this.getValue(object);
        if (!(value instanceof this.typeConstructor)) {
            return null;
        }
        const len = value.length;
        const persisting = new Array(len);
        const attachedState = Attribute.attachState(value);
        const persistedState = attachedState || [];
        let changed = !attachedState || attachedState.length !== len;
        const json = new Array(len);
        for (let i = 0; i < len; i += 1) {
            const el = value[i];
            json[i] = this.elementType.toJsonValue(state, el, options);
            persisting[i] = el;
            changed = changed || persistedState[i] !== el;
        }
        if (options.persisting) {
            Attribute.attachState(value, persisting, true);
        }
        if (changed) {
            state.setDirty();
        }
        return json;
    }
    /**
     * @inheritDoc
     */
    setJsonValue(state, object, json, options) {
        let value = null;
        if (json) {
            value = this.getValue(object);
            const len = json.length;
            if (!(value instanceof this.typeConstructor)) {
                value = new this.typeConstructor(len); // eslint-disable-line new-cap
            }
            const persisting = new Array(len);
            const persistedState = Attribute.attachState(value) || [];
            // clear additional items
            if (len < value.length) {
                value.splice(len, value.length - len);
            }
            for (let i = 0; i < len; i += 1) {
                const el = this.elementType.fromJsonValue(state, json[i], persistedState[i], options);
                value[i] = el;
                persisting[i] = el;
            }
            if (options.persisting) {
                Attribute.attachState(value, persisting, true);
            }
        }
        this.setValue(object, value);
    }
    /**
     * @inheritDoc
     */
    toJSON() {
        return Object.assign({ type: `${ListAttribute.ref}[${this.elementType.ref}]` }, super.toJSON());
    }
}
