// baqend/esm/lib/metamodel/MapAttribute.js
import { CollectionType, PluralAttribute } from './PluralAttribute';
import { Attribute } from './Attribute';
import { PersistentError } from '../error';
export class MapAttribute extends PluralAttribute {
    /**
     * @param name
     * @param keyType
     * @param elementType
     */
    constructor(name, keyType, elementType) {
        super(name, Map, elementType);
        this.keyType = keyType;
    }
    /**
     * Get the type id for this map type
     * @return
     */
    static get ref() {
        return '/db/collection.Map';
    }
    /**
     * @inheritDoc
     */
    get collectionType() {
        return CollectionType.MAP;
    }
    /**
     * @inheritDoc
     */
    getJsonValue(state, object, options) {
        const value = this.getValue(object);
        if (!(value instanceof this.typeConstructor)) {
            return null;
        }
        const persisting = {};
        const persistedState = Attribute.attachState(value) || {};
        let changed = Attribute.attachSize(value) !== value.size;
        const json = {};
        const iter = value.entries();
        for (let el = iter.next(); !el.done; el = iter.next()) {
            const entry = el.value;
            if (entry[0] === null || entry[0] === undefined) {
                throw new PersistentError('Map keys can\'t be null nor undefined.');
            }
            const jsonKey = this.keyValue(this.keyType.toJsonValue(state, entry[0], options));
            json[jsonKey] = this.elementType.toJsonValue(state, entry[1], options);
            persisting[jsonKey] = [entry[0], entry[1]];
            changed = changed || (persistedState[jsonKey] || [])[1] !== entry[1];
        }
        if (options.persisting) {
            Attribute.attachState(value, persisting, true);
            Attribute.attachSize(value, value.size);
        }
        if (changed) {
            state.setDirty();
        }
        return json;
    }
    /**
     * @inheritDoc
     */
    setJsonValue(state, object, json, options) {
        let value = null;
        if (json) {
            value = this.getValue(object);
            if (!(value instanceof this.typeConstructor)) {
                // eslint-disable-next-line new-cap
                value = new this.typeConstructor();
            }
            const persisting = {};
            const persistedState = Attribute.attachState(value) || {};
            value.clear();
            const jsonKeys = Object.keys(json);
            for (let i = 0, len = jsonKeys.length; i < len; i += 1) {
                const jsonKey = jsonKeys[i];
                const persistedEntry = persistedState[jsonKey] || [];
                // ensures that "false" keys will be converted to false, disallow null as keys
                const key = this.keyType.fromJsonValue(state, jsonKey, persistedEntry[0], options);
                const val = this.elementType.fromJsonValue(state, json[jsonKey], persistedEntry[1], options);
                persisting[jsonKey] = [key, val];
                value.set(key, val);
            }
            if (options.persisting) {
                Attribute.attachState(value, persisting, true);
                Attribute.attachSize(value, value.size);
            }
        }
        this.setValue(object, value);
    }
    /**
     * @inheritDoc
     */
    toJSON() {
        return Object.assign({ type: `${MapAttribute.ref}[${this.keyType.ref},${this.elementType.ref}]` }, super.toJSON());
    }
}
