// baqend/esm/lib/metamodel/SetAttribute.js
import { CollectionType, PluralAttribute } from './PluralAttribute';
import { Attribute } from './Attribute';
export class SetAttribute extends PluralAttribute {
    /**
     * Get the type id for this set type
     * @return
     */
    static get ref() {
        return '/db/collection.Set';
    }
    /**
     * @inheritDoc
     */
    get collectionType() {
        return CollectionType.SET;
    }
    /**
     * @param name The name of the attribute
     * @param elementType The element type of the collection
     */
    constructor(name, elementType) {
        super(name, Set, elementType);
    }
    /**
     * @inheritDoc
     */
    getJsonValue(state, object, options) {
        const value = this.getValue(object);
        if (!(value instanceof this.typeConstructor)) {
            return null;
        }
        const persisting = {};
        const persistedState = Attribute.attachState(value) || {};
        let changed = Attribute.attachSize(value) !== value.size;
        const json = [];
        const iter = value.values();
        for (let item = iter.next(); !item.done; item = iter.next()) {
            const el = item.value;
            const jsonValue = this.elementType.toJsonValue(state, el, options);
            json.push(jsonValue);
            const keyValue = this.keyValue(jsonValue);
            persisting[keyValue] = el;
            changed = changed || persistedState[keyValue] !== el;
        }
        if (options.persisting) {
            Attribute.attachState(value, persisting, true);
            Attribute.attachSize(value, value.size);
        }
        if (changed) {
            state.setDirty();
        }
        return json;
    }
    /**
     * @inheritDoc
     */
    setJsonValue(state, object, json, options) {
        let value = null;
        if (json) {
            value = this.getValue(object);
            if (!(value instanceof this.typeConstructor)) {
                value = new this.typeConstructor(); // eslint-disable-line new-cap
            }
            const persisting = {};
            const persistedState = Attribute.attachState(value) || {};
            value.clear();
            for (let i = 0, len = json.length; i < len; i += 1) {
                const jsonValue = json[i];
                const keyValue = this.keyValue(jsonValue);
                const el = this.elementType.fromJsonValue(state, jsonValue, persistedState[keyValue], options);
                value.add(el);
                persisting[keyValue] = el;
            }
            if (options.persisting) {
                Attribute.attachState(value, persisting, true);
                Attribute.attachSize(value, value.size);
            }
        }
        this.setValue(object, value);
    }
    /**
     * @inheritDoc
     */
    toJSON() {
        return Object.assign({ type: `${SetAttribute.ref}[${this.elementType.ref}]` }, super.toJSON());
    }
}
