// baqend/esm/lib/metamodel/DbIndex.js
/**
 * Creates a new index instance which is needed to create an
 * database index.
 */
export class DbIndex {
    /**
     * @param keys The name of the field which will be used
     * for the index,
     * an object of an field and index type combination or
     * an array of objects to create an compound index
     * @param unique Indicates if the index will be unique
     */
    constructor(keys, unique) {
        if (typeof keys === 'string') {
            const key = {};
            key[keys] = DbIndex.ASC;
            this.keys = [key];
        }
        else if (Array.isArray(keys)) {
            this.keys = keys;
        }
        else if (keys) {
            this.keys = [keys];
        }
        else {
            throw new Error('The keys parameter must be an String, Object or Array.');
        }
        this.unique = unique === true;
    }
    /**
     * Returns DbIndex Object created from the given JSON
     * @param json
     * @return
     */
    static fromJSON(json) {
        return new DbIndex(json.keys, json.unique);
    }
    /**
     * Indicates if this index is for the given field or includes it in a compound index
     * @param name The name of the field to check for
     * @return <code>true</code> if the index contains this field
     */
    hasKey(name) {
        for (let i = 0; i < this.keys.length; i += 1) {
            if (this.keys[i][name]) {
                return true;
            }
        }
        return false;
    }
    /**
     * Indicates if this index is a compound index of multiple attributes
     * @type boolean
     * @readonly
     */
    get isCompound() {
        return this.keys.length > 1;
    }
    /**
     * Indicates if this index is an unique index
     * @type boolean
     * @readonly
     */
    get isUnique() {
        return this.unique;
    }
    /**
     * Returns a JSON representation of the Index object
     *
     * @return A Json of this Index object
     */
    toJSON() {
        return {
            unique: this.unique,
            keys: this.keys,
        };
    }
}
DbIndex.ASC = 'asc';
DbIndex.DESC = 'desc';
