// baqend/esm/lib/metamodel/PluralAttribute.js
import { Attribute, PersistentAttributeType } from './Attribute';
export var CollectionType;
(function (CollectionType) {
    CollectionType[CollectionType["COLLECTION"] = 0] = "COLLECTION";
    CollectionType[CollectionType["LIST"] = 1] = "LIST";
    CollectionType[CollectionType["MAP"] = 2] = "MAP";
    CollectionType[CollectionType["SET"] = 3] = "SET";
})(CollectionType || (CollectionType = {}));
export class PluralAttribute extends Attribute {
    /**
     * @param name - The attribute name
     * @param typeConstructor - The collection constructor of the attribute
     * @param elementType - The type of the elements of the attribute collection
     */
    constructor(name, typeConstructor, elementType) {
        super(name);
        this.elementType = elementType;
        this.typeConstructor = typeConstructor;
    }
    /**
     * @inheritDoc
     */
    get persistentAttributeType() {
        return PersistentAttributeType.ELEMENT_COLLECTION;
    }
    /**
     * Retrieves a serialized string value of the given json which can be used as object keys
     * @param json The json of which the key should be retrieved
     * @return A serialized version of the json
     */
    keyValue(json) {
        if (json && typeof json === 'object' && 'id' in json) {
            return String(json.id);
        }
        return String(json);
    }
}
PluralAttribute.CollectionType = CollectionType;
