// baqend/esm/lib/metamodel/SingularAttribute.js
import { Attribute, PersistentAttributeType } from './Attribute';
import { PersistenceType } from './Type';
export class SingularAttribute extends Attribute {
    /**
     * @param name
     * @param type
     * @param isMetadata <code>true</code> if the attribute is an metadata attribute
     */
    constructor(name, type, isMetadata) {
        super(name, isMetadata);
        this.type = type;
    }
    /**
     * The constructor of the element type of this attribute
     */
    get typeConstructor() {
        return this.type.typeConstructor;
    }
    /**
     * @inheritDoc
     */
    get persistentAttributeType() {
        switch (this.type.persistenceType) {
            case PersistenceType.BASIC:
                return PersistentAttributeType.BASIC;
            case PersistenceType.EMBEDDABLE:
                return PersistentAttributeType.EMBEDDED;
            case PersistenceType.ENTITY:
                return PersistentAttributeType.ONE_TO_MANY;
            default:
                throw new Error('Unknown persistent attribute type.');
        }
    }
    /**
     * @inheritDoc
     */
    getJsonValue(state, object, options) {
        const persistedState = Attribute.attachState(object, {});
        const value = this.getValue(object);
        const changed = persistedState[this.name] !== value;
        if (options.persisting) {
            persistedState[this.name] = value;
        }
        if (changed) {
            state.setDirty();
        }
        return this.type.toJsonValue(state, value, options);
    }
    /**
     * @inheritDoc
     */
    setJsonValue(state, object, jsonValue, options) {
        const value = this.type.fromJsonValue(state, jsonValue, this.getValue(object), options);
        if (options.persisting) {
            const persistedState = Attribute.attachState(object, {});
            persistedState[this.name] = value;
        }
        this.setValue(object, value);
    }
    /**
     * @inheritDoc
     */
    toJSON() {
        return Object.assign({ type: this.type.ref }, super.toJSON());
    }
}
