// baqend/esm/lib/metamodel/Type.js
export var PersistenceType;
(function (PersistenceType) {
    PersistenceType[PersistenceType["BASIC"] = 0] = "BASIC";
    PersistenceType[PersistenceType["EMBEDDABLE"] = 1] = "EMBEDDABLE";
    PersistenceType[PersistenceType["ENTITY"] = 2] = "ENTITY";
    PersistenceType[PersistenceType["MAPPED_SUPERCLASS"] = 3] = "MAPPED_SUPERCLASS";
})(PersistenceType || (PersistenceType = {}));
export class Type {
    /**
     * @param ref
     * @param typeConstructor
     */
    constructor(ref, typeConstructor) {
        if (ref.indexOf('/db/') !== 0) {
            throw new SyntaxError(`Type ref ${ref} is invalid.`);
        }
        this.ref = ref;
        this.name = ref.substring(4);
        this._typeConstructor = typeConstructor;
    }
    /**
     * The persistent type of this type
     */
    get persistenceType() {
        return -1;
    }
    /**
     * @type boolean
     * @readonly
     */
    get isBasic() {
        return this.persistenceType === PersistenceType.BASIC;
    }
    /**
     * @type boolean
     * @readonly
     */
    get isEmbeddable() {
        return this.persistenceType === PersistenceType.EMBEDDABLE;
    }
    /**
     * @type boolean
     * @readonly
     */
    get isEntity() {
        return this.persistenceType === PersistenceType.ENTITY;
    }
    /**
     * @type boolean
     * @readonly
     */
    get isMappedSuperclass() {
        return this.persistenceType === PersistenceType.MAPPED_SUPERCLASS;
    }
    /**
     * The type constructor of this type
     */
    get typeConstructor() {
        return this._typeConstructor;
    }
    /**
     * @param typeConstructor - sets the type constructor of this type if it is not already set
     */
    set typeConstructor(typeConstructor) {
        if (this._typeConstructor) {
            throw new Error('typeConstructor has already been set.');
        }
        this._typeConstructor = typeConstructor;
    }
}
