// baqend/commonjs/lib/intersection/Logger.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
var msg = __importStar(require("../message"));
/**
 * A Logger to store log notes when running the app.
 */
var Logger = /** @class */ (function () {
    function Logger() {
        this.entityManager = null;
        this.levelIndex = 2;
    }
    /**
     * Creates a Logger instance for the given EntityManager
     * @param entityManager - Theo owning entityManager
     * @return The created logger instance
     */
    Logger.create = function (entityManager) {
        var proto = this.prototype;
        var logger = (function () {
            function LoggerFunction() {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                proto.log.apply(LoggerFunction, args);
            }
            Object.getOwnPropertyNames(proto).forEach(function (key) {
                Object.defineProperty(LoggerFunction, key, Object.getOwnPropertyDescriptor(proto, key));
            });
            return LoggerFunction;
        })();
        logger.init(entityManager);
        return logger;
    };
    Object.defineProperty(Logger.prototype, "level", {
        /**
         * The log level which will be logged
         *
         * The log level can be one of 'trace', 'debug', 'info', 'warn', 'error'
         * @type string
         */
        get: function () {
            return Logger.LEVELS[this.levelIndex];
        },
        /**
         * Sets the log level which will be logged
         * @param value
         */
        set: function (value) {
            var index = Logger.LEVELS.indexOf(value);
            if (index === -1) {
                throw new Error("Unknown logging level " + value);
            }
            this.levelIndex = index;
        },
        enumerable: false,
        configurable: true
    });
    Logger.prototype.log = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var level = Logger.LEVELS.indexOf(args[0]) === -1 ? 'info' : args.shift();
        if (this.levelIndex > Logger.LEVELS.indexOf(level)) {
            return Promise.resolve(null);
        }
        var message = typeof args[0] === 'string' ? this.format(args.shift(), args) : '[no message]';
        var data = null;
        if (args.length) {
            var arg = args.pop();
            data = arg;
            if (typeof arg !== 'object' || Array.isArray(arg)) {
                data = { data: arg };
            }
            if (arg instanceof Error) {
                // errors aren't loggable by default, since they do not have any visible property
                var 
                // @ts-ignore
                stack = arg.stack, data1 = arg.data, message1 = arg.message, name_1 = arg.name, status_1 = arg.status;
                data = {
                    name: name_1,
                    message: message1,
                    stack: stack,
                    status: status_1,
                    data: data1,
                };
            }
        }
        if (args.length) {
            message += ", " + args.join(', ');
        }
        return this.logJSON(__assign({ date: new Date().toISOString(), message: message,
            level: level,
            data: data }, (this.entityManager.me && { user: this.entityManager.me.id })));
    };
    Logger.prototype.format = function (message, args) {
        if (args.length === 0) {
            return message;
        }
        var str = String(message).replace(Logger.FORMAT_REGEXP, function (x) {
            if (x === '%%') {
                return '%';
            }
            if (!args.length) {
                return x;
            }
            switch (x) {
                case '%s':
                    return String(args.shift());
                case '%d':
                    return String(Number(args.shift()));
                case '%j':
                    try {
                        return JSON.stringify(args.shift());
                    }
                    catch (_) {
                        return '[Circular]';
                    }
                default:
                    return x;
            }
        });
        return str;
    };
    Logger.prototype.init = function (entityManager) {
        var _this = this;
        this.entityManager = entityManager;
        this.levelIndex = 2;
        Logger.LEVELS.forEach(function (level) {
            _this[level] = _this.log.bind(_this, level);
        });
    };
    Logger.prototype.logJSON = function (json) {
        if (!this.entityManager.isReady) {
            return this.entityManager.ready(this.logJSON.bind(this, json));
        }
        return this.entityManager.send(new msg.CreateObject('logs.AppLog', json));
    };
    Logger.LEVELS = ['trace', 'debug', 'info', 'warn', 'error'];
    Logger.FORMAT_REGEXP = /%[sdj%]/g;
    return Logger;
}());
exports.Logger = Logger;
