// baqend/commonjs/lib/query/Builder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
var Filter_1 = require("./Filter");
var Condition_1 = require("./Condition");
var Operator_1 = require("./Operator");
var Query_1 = require("./Query");
var Node_1 = require("./Node");
var Builder = /** @class */ (function (_super) {
    __extends(Builder, _super);
    function Builder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Joins the conditions by an logical AND
     * @param args The query nodes to join
     * @return Returns a new query which joins the given queries by a logical AND
     */
    Builder.prototype.and = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this.addOperator('$and', Query_1.flatArgs(args));
    };
    /**
     * Joins the conditions by an logical OR
     * @param args The query nodes to join
     * @return Returns a new query which joins the given queries by a logical OR
     */
    Builder.prototype.or = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this.addOperator('$or', Query_1.flatArgs(args));
    };
    /**
     * Joins the conditions by an logical NOR
     * @param args The query nodes to join
     * @return Returns a new query which joins the given queries by a logical NOR
     */
    Builder.prototype.nor = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this.addOperator('$nor', Query_1.flatArgs(args));
    };
    Builder.prototype.eventStream = function (options, onNext, onError, onComplete) {
        return this.where({}).eventStream(options, onNext, onError, onComplete);
    };
    Builder.prototype.resultStream = function (options, onNext, onError, onComplete) {
        return this.where({}).resultStream(options, onNext, onError, onComplete);
    };
    /**
     * @inheritDoc
     */
    Builder.prototype.resultList = function (options, doneCallback, failCallback) {
        return this.where({}).resultList(options, doneCallback, failCallback);
    };
    /**
     * @inheritDoc
     */
    Builder.prototype.singleResult = function (options, doneCallback, failCallback) {
        return this.where({}).singleResult(options, doneCallback, failCallback);
    };
    /**
     * @inheritDoc
     */
    Builder.prototype.count = function (doneCallback, failCallback) {
        return this.where({}).count(doneCallback, failCallback);
    };
    Builder.prototype.addOperator = function (operator, args) {
        if (args.length < 2) {
            throw new Error("Only two or more queries can be joined with an " + operator + " operator.");
        }
        args.forEach(function (arg, index) {
            if (!(arg instanceof Node_1.Node)) {
                throw new Error("Argument at index " + index + " is not a query.");
            }
        });
        return new Operator_1.Operator(this.entityManager, this.resultClass, operator, args);
    };
    Builder.prototype.addOrder = function (fieldOrSort, order) {
        return new Filter_1.Filter(this.entityManager, this.resultClass).addOrder(fieldOrSort, order);
    };
    Builder.prototype.addFilter = function (field, filter, value) {
        return new Filter_1.Filter(this.entityManager, this.resultClass).addFilter(field, filter, value);
    };
    Builder.prototype.addOffset = function (offset) {
        return new Filter_1.Filter(this.entityManager, this.resultClass).addOffset(offset);
    };
    Builder.prototype.addLimit = function (limit) {
        return new Filter_1.Filter(this.entityManager, this.resultClass).addLimit(limit);
    };
    return Builder;
}(Query_1.Query));
exports.Builder = Builder;
Object.assign(Builder.prototype, Condition_1.Condition);
