// recursive-rename/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _traverser = _interopRequireDefault(require("./traverser"));

var _validator = _interopRequireDefault(require("./validator"));

var _excludes = _interopRequireDefault(require("./excludes"));

var Rename = /*#__PURE__*/function () {
  function Rename(options) {
    (0, _classCallCheck2["default"])(this, Rename);
    var src = options.src,
        dest = options.dest,
        exclude = options.exclude,
        override = options.override,
        path = options.path;

    var _Excludes = new _excludes["default"](exclude),
        excludes = _Excludes.excludes;

    var validatorOptions = {
      excludes: excludes,
      override: override,
      path: path
    };
    this.validator = new _validator["default"](src, dest, validatorOptions);
  }

  (0, _createClass2["default"])(Rename, [{
    key: "dive",
    value: function () {
      var _dive = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(dry) {
        var _this$validator, src, dest, excludes, path, traverser;

        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$validator = this.validator, src = _this$validator.src, dest = _this$validator.dest, excludes = _this$validator.excludes, path = _this$validator.path;
                traverser = new _traverser["default"](path, {
                  src: src,
                  dest: dest,
                  excludes: excludes
                });
                _context.next = 4;
                return traverser.traverse({
                  dry: dry
                });

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function dive(_x) {
        return _dive.apply(this, arguments);
      }

      return dive;
    }()
  }]);
  return Rename;
}();

module.exports = Rename;
var _default = Rename;
exports["default"] = _default;
