// baqend/commonjs/lib/binding/Factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Factory = void 0;
// @ts-ignore
var Factory = /** @class */ (function () {
    function Factory() {
        this.type = null;
        this.prototype = null;
    }
    Factory.extend = function (target, proto) {
        if (proto !== Factory.prototype) {
            this.extend(target, Object.getPrototypeOf(proto));
        }
        var properties = Object.getOwnPropertyNames(proto);
        for (var j = 0, len = properties.length; j < len; j += 1) {
            var prop = properties[j];
            Object.defineProperty(target, prop, Object.getOwnPropertyDescriptor(proto, prop));
        }
        return target;
    };
    /**
     * Creates a new Factory for the given type
     * @param type - the type constructor of T
     * @return A new object factory to created instances of T
     */
    Factory.createFactory = function (type) {
        // We want te explicitly name the created factory and give the constructor a properly argument name
        var factory = Factory.extend((function FactoryConstructor() {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return factory.newInstance(args);
        }), this.prototype);
        // lets instanceof work properly
        factory.prototype = type.prototype;
        factory.type = type;
        return factory;
    };
    /**
     * Creates a new instance of the factory type
     * @param args Constructor arguments used for instantiation
     * @return A new created instance of *
     * @instance
     */
    Factory.prototype.new = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this.newInstance(args);
    };
    /**
     * Creates a new instance of the factory type
     * @param args Constructor arguments used for instantiation
     * @return A new created instance of *
     * @instance
     */
    Factory.prototype.newInstance = function (args) {
        if (!args || args.length === 0) {
            // eslint-disable-next-line new-cap
            return new this.type();
        }
        // es6 constructors can't be called, therefore bind all arguments and invoke the constructor
        // then with the bounded parameters
        // The first argument is shift out by invocation with `new`.
        var a = [null];
        Array.prototype.push.apply(a, args);
        var boundConstructor = (Function.prototype.bind.apply(this.type, a));
        // eslint-disable-next-line new-cap
        return new boundConstructor();
    };
    return Factory;
}());
exports.Factory = Factory;
