// baqend/commonjs/lib/metamodel/Type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Type = exports.PersistenceType = void 0;
var PersistenceType;
(function (PersistenceType) {
    PersistenceType[PersistenceType["BASIC"] = 0] = "BASIC";
    PersistenceType[PersistenceType["EMBEDDABLE"] = 1] = "EMBEDDABLE";
    PersistenceType[PersistenceType["ENTITY"] = 2] = "ENTITY";
    PersistenceType[PersistenceType["MAPPED_SUPERCLASS"] = 3] = "MAPPED_SUPERCLASS";
})(PersistenceType = exports.PersistenceType || (exports.PersistenceType = {}));
var Type = /** @class */ (function () {
    /**
     * @param ref
     * @param typeConstructor
     */
    function Type(ref, typeConstructor) {
        if (ref.indexOf('/db/') !== 0) {
            throw new SyntaxError("Type ref " + ref + " is invalid.");
        }
        this.ref = ref;
        this.name = ref.substring(4);
        this._typeConstructor = typeConstructor;
    }
    Object.defineProperty(Type.prototype, "persistenceType", {
        /**
         * The persistent type of this type
         */
        get: function () {
            return -1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Type.prototype, "isBasic", {
        /**
         * @type boolean
         * @readonly
         */
        get: function () {
            return this.persistenceType === PersistenceType.BASIC;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Type.prototype, "isEmbeddable", {
        /**
         * @type boolean
         * @readonly
         */
        get: function () {
            return this.persistenceType === PersistenceType.EMBEDDABLE;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Type.prototype, "isEntity", {
        /**
         * @type boolean
         * @readonly
         */
        get: function () {
            return this.persistenceType === PersistenceType.ENTITY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Type.prototype, "isMappedSuperclass", {
        /**
         * @type boolean
         * @readonly
         */
        get: function () {
            return this.persistenceType === PersistenceType.MAPPED_SUPERCLASS;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Type.prototype, "typeConstructor", {
        /**
         * The type constructor of this type
         */
        get: function () {
            return this._typeConstructor;
        },
        /**
         * @param typeConstructor - sets the type constructor of this type if it is not already set
         */
        set: function (typeConstructor) {
            if (this._typeConstructor) {
                throw new Error('typeConstructor has already been set.');
            }
            this._typeConstructor = typeConstructor;
        },
        enumerable: false,
        configurable: true
    });
    return Type;
}());
exports.Type = Type;
