// baqend/esm/lib/binding/DeviceFactory.js
import * as message from '../message';
import { PushMessage } from '../intersection';
import { EntityFactory } from './EntityFactory';
export class DeviceFactory extends EntityFactory {
    /**
     * Push message will be used to send a push notification to a set of devices
     */
    get PushMessage() {
        return PushMessage;
    }
    /**
     * The current registered device, or <code>null</code> if the device is not registered
     * @type model.Device
     */
    get me() {
        return this.db.deviceMe;
    }
    /**
     * Returns true if the devices is already registered, otherwise false.
     * @type boolean
     */
    get isRegistered() {
        return this.db.isDeviceRegistered;
    }
    /**
     * Loads the Public VAPID Key which can be used to subscribe a Browser for Web Push notifications
     * @return The public VAPID Web Push subscription key
     */
    loadWebPushKey() {
        const msg = new message.VAPIDPublicKey();
        msg.responseType('arraybuffer');
        return this.db.send(msg).then((response) => response.entity);
    }
    register(os, tokenOrSubscription, device, doneCallback, failCallback) {
        if (device instanceof Function) {
            return this.register(os, tokenOrSubscription, null, device, doneCallback);
        }
        const subscription = typeof tokenOrSubscription === 'string' ? { token: tokenOrSubscription } : tokenOrSubscription;
        return this.db.registerDevice(os, subscription, device).then(doneCallback, failCallback);
    }
    /**
     * Uses the info from the given {@link PushMessage} message to send an push notification.
     * @param pushMessage to send an push notification.
     * @param doneCallback Called when the operation succeed.
     * @param failCallback Called when the operation failed.
     * @return
     */
    push(pushMessage, doneCallback, failCallback) {
        return this.db.pushDevice(pushMessage).then(doneCallback, failCallback);
    }
}
