// baqend/commonjs/lib/connector/Connector.js
"use strict";
/* eslint-disable no-restricted-globals */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connector = void 0;
var error_1 = require("../error");
var Connector = /** @class */ (function () {
    /**
     * @param host - the host to connect to
     * @param port - the port to connect to
     * @param secure - <code>true</code> for an secure connection
     * @param basePath - The base path of the api endpoint
     */
    function Connector(host, port, secure, basePath) {
        this.host = host;
        this.port = port;
        this.secure = secure;
        this.basePath = basePath;
        /**
         * the origin do not contains the base path
         */
        this.origin = Connector.toUri(this.host, this.port, this.secure, '');
        /**
         * The connector will detect if gzip is supported.
         * Returns true if supported otherwise false.
         */
        this.gzip = false;
    }
    /**
     * Indicates id this connector is usable in the current runtime environment
     * This method must be overwritten in subclass implementations
     * @param host - the host to connect to
     * @param port - the port to connect to
     * @param secure - <code>true</code> for an secure connection
     * @param basePath - The base path of the api endpoint
     * @return <code>true</code> if this connector is usable in the current environment
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    Connector.isUsable = function (host, port, secure, basePath) {
        return false;
    };
    /**
     * @param host or location
     * @param port
     * @param secure=true <code>true</code> for an secure connection
     * @param basePath The basepath of the api
     * @return
     */
    Connector.create = function (host, port, secure, basePath) {
        var h = host;
        var p = port;
        var s = secure;
        var b = basePath;
        if (typeof location !== 'undefined') {
            if (!h) {
                h = location.hostname;
                p = Number(location.port);
            }
            if (s === undefined) {
                s = location.protocol === 'https:';
            }
        }
        // ensure right type, make secure: true the default
        s = s === undefined || !!s;
        if (b === undefined) {
            b = Connector.DEFAULT_BASE_PATH;
        }
        if (h.indexOf('/') !== -1) {
            var matches = /^(https?):\/\/([^/:]+|\[[^\]]+])(:(\d*))?(\/\w+)?\/?$/.exec(h);
            if (matches) {
                s = matches[1] === 'https';
                h = matches[2].replace(/(\[|])/g, '');
                p = Number(matches[4]);
                b = matches[5] || '';
            }
            else {
                throw new Error("The connection uri host " + h + " seems not to be valid");
            }
        }
        else if (h !== 'localhost' && /^[a-z0-9-]*$/.test(h)) {
            // handle app names as hostname
            h += Connector.HTTP_DOMAIN;
        }
        if (!p) {
            p = s ? 443 : 80;
        }
        var url = Connector.toUri(h, p, s, b);
        var connection = this.connections[url];
        if (!connection) {
            // check last registered connector first to simplify registering connectors
            for (var i = this.connectors.length - 1; i >= 0; i -= 1) {
                var ConnectorConstructor = this.connectors[i];
                if (ConnectorConstructor.isUsable && ConnectorConstructor.isUsable(h, p, s, b)) {
                    connection = new ConnectorConstructor(h, p, s, b);
                    break;
                }
            }
            if (!connection) {
                throw new Error('No connector is usable for the requested connection.');
            }
            this.connections[url] = connection;
        }
        return connection;
    };
    Connector.toUri = function (host, port, secure, basePath) {
        var uri = (secure ? 'https://' : 'http://') + (host.indexOf(':') !== -1 ? "[" + host + "]" : host);
        uri += ((secure && port !== 443) || (!secure && port !== 80)) ? ":" + port : '';
        uri += basePath;
        return uri;
    };
    /**
     * @param message
     * @return
     */
    Connector.prototype.send = function (message) {
        var _this = this;
        var response = { status: 0, headers: {} };
        return Promise.resolve()
            .then(function () { return _this.prepareRequest(message); })
            .then(function () { return new Promise(function (resolve) {
            _this.doSend(message, message.request, resolve);
        }); })
            .then(function (res) { response = res; })
            .then(function () { return _this.prepareResponse(message, response); })
            .then(function () {
            message.doReceive(response);
            return response;
        })
            .catch(function (e) {
            response.entity = null;
            throw error_1.PersistentError.of(e);
        });
    };
    /**
     * @param message
     * @return
     */
    Connector.prototype.prepareRequest = function (message) {
        var _this = this;
        var mimeType = message.mimeType();
        if (!mimeType) {
            var type = message.request.type;
            if (type === 'json') {
                message.mimeType('application/json;charset=utf-8');
            }
            else if (type === 'text') {
                message.mimeType('text/plain;charset=utf-8');
            }
        }
        this.toFormat(message);
        var accept;
        switch (message.responseType()) {
            case 'json':
                accept = 'application/json';
                break;
            case 'text':
                accept = 'text/*';
                break;
            default:
                accept = 'application/json,text/*;q=0.5,*/*;q=0.1';
        }
        if (!message.accept()) {
            message.accept(accept);
        }
        if (this.gzip) {
            var ifNoneMatch = message.ifNoneMatch();
            if (ifNoneMatch && ifNoneMatch !== '""' && ifNoneMatch !== '*') {
                message.ifNoneMatch(ifNoneMatch.slice(0, -1) + "--gzip\"");
            }
        }
        var tokenStorage = message.tokenStorage();
        if (message.request.path === '/connect') {
            return tokenStorage.signPath(this.basePath + message.request.path)
                .then(function (signedPath) {
                // eslint-disable-next-line no-param-reassign
                message.request.path = signedPath.substring(_this.basePath.length);
                if (message.cacheControl()) {
                    // eslint-disable-next-line no-param-reassign
                    message.request.path += (message.request.path.indexOf('?') !== -1 ? '&' : '?') + "BCB";
                }
                return message;
            });
        }
        if (tokenStorage) {
            var token = tokenStorage.token;
            if (token) {
                message.header('authorization', "BAT " + token);
            }
        }
        return message;
    };
    /**
     * @param message
     * @param response The received response headers and data
     * @return
     */
    Connector.prototype.prepareResponse = function (message, response) {
        var _this = this;
        // IE9 returns status code 1223 instead of 204
        response.status = response.status === 1223 ? 204 : response.status;
        var type;
        var headers = response.headers || {};
        // some proxies send content back on 204 responses
        var entity = response.status === 204 ? null : response.entity;
        if (entity) {
            type = message.responseType();
            if (!type || response.status >= 400) {
                var contentType = headers['content-type'] || headers['Content-Type'];
                if (contentType && contentType.indexOf('application/json') > -1) {
                    type = 'json';
                }
            }
        }
        if (headers.etag) {
            headers.etag = headers.etag.replace('--gzip', '');
        }
        var tokenStorage = message.tokenStorage();
        if (tokenStorage) {
            var token = headers['baqend-authorization-token'] || headers['Baqend-Authorization-Token'];
            if (token) {
                tokenStorage.update(token);
            }
        }
        return new Promise(function (resolve) {
            resolve(entity && _this.fromFormat(response, entity, type));
        }).then(function (resultEntity) {
            response.entity = resultEntity;
            if (message.request.path.indexOf('/connect') !== -1 && resultEntity) {
                _this.gzip = !!resultEntity.gzip;
            }
        }, function (e) {
            throw new Error("Response was not valid " + type + ": " + e.message);
        });
    };
    Connector.DEFAULT_BASE_PATH = '/v1';
    Connector.HTTP_DOMAIN = '.app.baqend.com';
    /**
     * An array of all exposed response headers
     */
    Connector.RESPONSE_HEADERS = [
        'baqend-authorization-token',
        'content-type',
        'baqend-size',
        'baqend-acl',
        'etag',
        'last-modified',
        'baqend-created-at',
        'baqend-custom-headers',
    ];
    /**
     * Array of all available connector implementations
     */
    Connector.connectors = [];
    /**
     * Array of all created connections
     */
    Connector.connections = {};
    return Connector;
}());
exports.Connector = Connector;
