// baqend/commonjs/lib/connector/XMLHttpConnector.js
"use strict";
/* this connector will only be choose in browser compatible environments */
/* eslint no-restricted-globals: ["off", "addEventListener", "removeEventListener"] */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMLHttpConnector = void 0;
var Connector_1 = require("./Connector");
var util_1 = require("../util");
var XMLHttpConnector = /** @class */ (function (_super) {
    __extends(XMLHttpConnector, _super);
    function XMLHttpConnector() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @inheritDoc
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    XMLHttpConnector.isUsable = function (host, port, secure, basePath) {
        return typeof XMLHttpRequest !== 'undefined';
    };
    /**
     * @inheritDoc
     */
    XMLHttpConnector.prototype.doSend = function (message, request, receive) {
        var _this = this;
        if (request.method === 'OAUTH') {
            if (this.oAuthHandle) {
                this.oAuthHandle({ status: 409, headers: {}, entity: '{"message": "A new OAuth request was sent."}' });
            }
            localStorage.removeItem('oauth-response');
            var handler_1 = function (event) {
                if (event.key === 'oauth-response' && _this.oAuthHandle && event.newValue) {
                    _this.oAuthHandle(JSON.parse(event.newValue));
                }
            };
            this.oAuthHandle = function (msg) {
                receive(msg);
                localStorage.removeItem('oauth-response');
                removeEventListener('storage', handler_1, false);
            };
            addEventListener('storage', handler_1, false);
            return;
        }
        var xhr = new XMLHttpRequest();
        var url = this.origin + this.basePath + request.path;
        xhr.onreadystatechange = function () {
            // if we receive an error switch the response type to json
            if (xhr.responseType && xhr.readyState === 2 && xhr.status >= 400) {
                xhr.responseType = 'text';
            }
            if (xhr.readyState === 4) {
                var response_1 = {
                    headers: {},
                    status: xhr.status,
                    entity: xhr.response || xhr.responseText,
                };
                Connector_1.Connector.RESPONSE_HEADERS.forEach(function (name) {
                    response_1.headers[name] = xhr.getResponseHeader(name) || '';
                });
                receive(response_1);
            }
        };
        // Set the message progress callback
        if (xhr.upload && message.progress()) {
            xhr.upload.onprogress = message.progress();
        }
        xhr.open(request.method, url, true);
        var entity = request.entity;
        var headers = request.headers;
        var headerNames = Object.keys(headers);
        for (var i = 0, len = headerNames.length; i < len; i += 1) {
            var headerName = headerNames[i];
            xhr.setRequestHeader(headerName, headers[headerName]);
        }
        xhr.withCredentials = message.withCredentials;
        switch (message.responseType()) {
            case 'arraybuffer':
                xhr.responseType = 'arraybuffer';
                break;
            case 'blob':
            case 'data-url':
            case 'base64':
                xhr.responseType = 'blob';
                break;
            default:
            // ignore
        }
        xhr.send(entity);
    };
    /**
     * @inheritDoc
     */
    XMLHttpConnector.prototype.fromFormat = function (response, entity, type) {
        if (type === 'json') {
            return JSON.parse(entity);
        }
        if (type === 'data-url' || type === 'base64') {
            var reader_1 = new FileReader();
            reader_1.readAsDataURL(entity);
            return new Promise(function (resolve, reject) {
                reader_1.onload = resolve;
                reader_1.onerror = reject;
            }).then(function () {
                var result = reader_1.result;
                if (type === 'base64' && typeof result === 'string') {
                    result = result.substring(result.indexOf(',') + 1);
                }
                return result;
            });
        }
        return entity;
    };
    /**
     * @inheritDoc
     */
    XMLHttpConnector.prototype.toFormat = function (message) {
        var type = message.request.type;
        if (type) {
            var entity = message.request.entity;
            var mimeType = message.mimeType();
            switch (type) {
                case 'blob':
                    mimeType = mimeType || entity.type;
                    break;
                case 'arraybuffer':
                case 'form':
                    break;
                case 'data-url': {
                    var match = entity.match(/^data:(.+?)(;base64)?,(.*)$/);
                    var isBase64 = match[2];
                    // eslint-disable-next-line prefer-destructuring
                    entity = match[3];
                    type = 'blob';
                    mimeType = mimeType || match[1];
                    if (!isBase64) {
                        entity = decodeURIComponent(entity);
                        break;
                    }
                }
                // fallthrough
                case 'base64': {
                    var binaryStr = util_1.atob(entity);
                    var len = binaryStr.length;
                    var array = new Uint8Array(len);
                    for (var i = 0; i < len; i += 1) {
                        array[i] = binaryStr.charCodeAt(i);
                    }
                    type = 'blob';
                    entity = new Blob([array], { type: mimeType });
                    break;
                }
                case 'json':
                    if (typeof entity !== 'string') {
                        entity = JSON.stringify(entity);
                    }
                    break;
                case 'text':
                    break;
                default:
                    throw new Error("Supported request format:" + type);
            }
            message.entity(entity, type).mimeType(mimeType);
        }
    };
    return XMLHttpConnector;
}(Connector_1.Connector));
exports.XMLHttpConnector = XMLHttpConnector;
Connector_1.Connector.connectors.push(XMLHttpConnector);
