// baqend/esm/lib/intersection/Code.js
import * as message from '../message';
import { StatusCode } from '../connector';
import { Validator } from './Validator';
/**
 * Representation of a Code which runs on Baqend.
 */
export class Code {
    /**
     * @param metamodel
     * @param entityManagerFactory
     */
    constructor(metamodel, entityManagerFactory) {
        this.metamodel = metamodel;
        this.entityManagerFactory = entityManagerFactory;
    }
    /**
     * Converts the given function to a string
     * @param fn The JavaScript function to serialize
     * @return The serialized function
     */
    functionToString(fn) {
        if (!fn) {
            return '';
        }
        let str = fn.toString();
        str = str.substring(str.indexOf('{') + 1, str.lastIndexOf('}'));
        if (str.charAt(0) === '\n') {
            str = str.substring(1);
        }
        if (str.charAt(str.length - 1) === '\n') {
            str = str.substring(0, str.length - 1);
        }
        return str;
    }
    /**
     * Converts the given string to a module wrapper function
     * @param signature The expected parameters of the function
     * @param code The JavaScript function to deserialize
     * @return The deserialized function
     */
    stringToFunction(signature, code) {
        // eslint-disable-next-line @typescript-eslint/no-implied-eval
        return new Function(signature /* typings are incorrect here */, code); // eslint-disable-line no-new-func
    }
    /**
     * Loads a list of all available modules without handlers
     *
     * @return
     */
    loadModules() {
        const msg = new message.GetAllModules();
        return this.entityManagerFactory.send(msg)
            .then((response) => response.entity);
    }
    loadCode(type, codeType, asFunction = false) {
        const bucket = typeof type === 'string' ? type : type.name;
        const msg = new message.GetBaqendCode(bucket, codeType)
            .responseType('text');
        return this.entityManagerFactory.send(msg)
            .then((response) => this.parseCode(bucket, codeType, asFunction, response.entity), (e) => {
            if (e.status === StatusCode.OBJECT_NOT_FOUND) {
                return null;
            }
            throw e;
        });
    }
    saveCode(type, codeType, fn) {
        const bucket = typeof type === 'string' ? type : type.name;
        const msg = new message.SetBaqendCode(bucket, codeType)
            .entity(fn instanceof Function ? this.functionToString(fn) : fn, 'text')
            .responseType('text');
        return this.entityManagerFactory.send(msg)
            .then((response) => this.parseCode(bucket, codeType, fn instanceof Function, response.entity));
    }
    /**
     * Deletes Baqend code identified by the given bucket and code type
     *
     * @param type The entity type for the handler or the Name of the
     * Baqend code
     * @param codeType The type of the code
     * @return succeed if the code was deleted
     */
    deleteCode(type, codeType) {
        const bucket = typeof type === 'string' ? type : type.name;
        const msg = new message.DeleteBaqendCode(bucket, codeType);
        return this.entityManagerFactory.send(msg)
            .then(() => this.parseCode(bucket, codeType, false, null));
    }
    /**
     * @param bucket
     * @param codeType
     * @param [asFunction=false]
     * @param code
     * @return
     * @private
     */
    parseCode(bucket, codeType, asFunction, code) {
        if (codeType === 'validate') {
            const type = this.metamodel.entity(bucket);
            type.validationCode = code === null ? null : Validator.compile(type, code);
            return asFunction ? type.validationCode : code;
        }
        return code && asFunction ? this.stringToFunction(['module', 'exports'], code) : code;
    }
}
