// baqend/esm/lib/intersection/Logger.js
import * as msg from '../message';
/**
 * A Logger to store log notes when running the app.
 */
export class Logger {
    constructor() {
        this.entityManager = null;
        this.levelIndex = 2;
    }
    /**
     * Creates a Logger instance for the given EntityManager
     * @param entityManager - Theo owning entityManager
     * @return The created logger instance
     */
    static create(entityManager) {
        const proto = this.prototype;
        const logger = (() => {
            function LoggerFunction(...args) {
                proto.log.apply(LoggerFunction, args);
            }
            Object.getOwnPropertyNames(proto).forEach((key) => {
                Object.defineProperty(LoggerFunction, key, Object.getOwnPropertyDescriptor(proto, key));
            });
            return LoggerFunction;
        })();
        logger.init(entityManager);
        return logger;
    }
    /**
     * The log level which will be logged
     *
     * The log level can be one of 'trace', 'debug', 'info', 'warn', 'error'
     * @type string
     */
    get level() {
        return Logger.LEVELS[this.levelIndex];
    }
    /**
     * Sets the log level which will be logged
     * @param value
     */
    set level(value) {
        const index = Logger.LEVELS.indexOf(value);
        if (index === -1) {
            throw new Error(`Unknown logging level ${value}`);
        }
        this.levelIndex = index;
    }
    log(...args) {
        const level = Logger.LEVELS.indexOf(args[0]) === -1 ? 'info' : args.shift();
        if (this.levelIndex > Logger.LEVELS.indexOf(level)) {
            return Promise.resolve(null);
        }
        let message = typeof args[0] === 'string' ? this.format(args.shift(), args) : '[no message]';
        let data = null;
        if (args.length) {
            const arg = args.pop();
            data = arg;
            if (typeof arg !== 'object' || Array.isArray(arg)) {
                data = { data: arg };
            }
            if (arg instanceof Error) {
                // errors aren't loggable by default, since they do not have any visible property
                const { 
                // @ts-ignore
                stack, data: data1, message: message1, name, status, } = arg;
                data = {
                    name,
                    message: message1,
                    stack,
                    status,
                    data: data1,
                };
            }
        }
        if (args.length) {
            message += `, ${args.join(', ')}`;
        }
        return this.logJSON(Object.assign({ date: new Date().toISOString(), message,
            level,
            data }, (this.entityManager.me && { user: this.entityManager.me.id })));
    }
    format(message, args) {
        if (args.length === 0) {
            return message;
        }
        const str = String(message).replace(Logger.FORMAT_REGEXP, (x) => {
            if (x === '%%') {
                return '%';
            }
            if (!args.length) {
                return x;
            }
            switch (x) {
                case '%s':
                    return String(args.shift());
                case '%d':
                    return String(Number(args.shift()));
                case '%j':
                    try {
                        return JSON.stringify(args.shift());
                    }
                    catch (_) {
                        return '[Circular]';
                    }
                default:
                    return x;
            }
        });
        return str;
    }
    init(entityManager) {
        this.entityManager = entityManager;
        this.levelIndex = 2;
        Logger.LEVELS.forEach((level) => {
            this[level] = this.log.bind(this, level);
        });
    }
    logJSON(json) {
        if (!this.entityManager.isReady) {
            return this.entityManager.ready(this.logJSON.bind(this, json));
        }
        return this.entityManager.send(new msg.CreateObject('logs.AppLog', json));
    }
}
Logger.LEVELS = ['trace', 'debug', 'info', 'warn', 'error'];
Logger.FORMAT_REGEXP = /%[sdj%]/g;
