// baqend/esm/lib/intersection/PushMessage.js
import { Entity } from '../binding';
/**
 * PushMessages are used to send a push notification to a set of devices
 */
export class PushMessage {
    /**
     * Push message will be used to send a push notification to a set of devices
     *
     * @param [devices] The Set of device references which
     * will receive this push notification.
     * @param message The message of the push notification.
     * @param subject The subject of the push notification.
     * @param [options] The options object which can contain additional information and data
     * @param [badge] The badge for iOS or Web Push devices
     * @param [data] The data object which can contain additional information.
     */
    constructor(devices, message, subject, options, badge, data) {
        const opts = typeof options === 'string' ? { sound: options, badge, data } : (options || {});
        this.devices = PushMessage.initDevices(devices);
        this.message = message;
        this.subject = subject;
        Object.assign(this, opts);
    }
    /**
     * Instantiates a set of devices from the given parameter
     * @param devices
     * @return
     */
    static initDevices(devices) {
        if (devices instanceof Set) {
            return devices;
        }
        if (devices instanceof Entity) {
            return new Set([devices]);
        }
        if (!devices || devices[Symbol.iterator]) {
            return new Set(devices);
        }
        throw new Error('Only Sets, Lists and Arrays can be used as devices.');
    }
    /**
     * Adds a new object to the set of devices
     * @param device will be added to the device set to receive the push notification
     * @return
     */
    addDevice(device) {
        this.devices.add(device);
    }
    /**
     * Converts the push message to JSON
     * @return
     */
    toJSON() {
        if (!this.devices || !this.devices.size) {
            throw new Error('Set of devices is empty.');
        }
        return Object.assign({}, this, {
            devices: Array.from(this.devices, (device) => device.id),
        });
    }
}
