// baqend/esm/lib/intersection/Validator.js
const FallBachValLib = {};
let valLib = FallBachValLib;
try {
    // we load this module as an optional external dependency
    // eslint-disable-next-line global-require
    valLib = require('validator');
}
catch (e) {
    // ignore loading optional module error
}
export class Validator {
    constructor(key, entity) {
        /**
         * The cached errors of the validation
         */
        this.errors = [];
        this.key = key;
        this.entity = entity;
    }
    /**
     * Compiles the given validation code for the managedType
     * @param managedType The managedType of the code
     * @param validationCode The validation code
     * @return the parsed validation function
     */
    static compile(managedType, validationCode) {
        const keys = [];
        const iter = managedType.attributes();
        for (let el = iter.next(); !el.done; el = iter.next()) {
            const attr = el.value;
            keys.push(attr.name);
        }
        // eslint-disable-next-line @typescript-eslint/no-implied-eval,no-new-func
        const fn = new Function(...keys, validationCode);
        return function onValidate(argObj) {
            if (valLib === FallBachValLib) {
                throw new Error('Validation code will not be executed. Make sure that the validator package is correctly provided as an external dependency.');
            }
            const args = keys.map((name) => argObj[name]);
            return fn.apply({}, args);
        };
    }
    /**
     * Gets the value of the attribute
     * @return Value
     */
    get value() {
        return this.entity[this.key];
    }
    /**
     * Checks if the attribute is valid
     * @return
     */
    get isValid() {
        return this.errors.length === 0;
    }
    is(error, fn) {
        if (error instanceof Function) {
            return this.is('is', error);
        }
        if (fn(this.value, valLib) === false) {
            this.errors.push(error);
        }
        return this;
    }
    callMethod(method, errorMessage, argumentList) {
        const args = argumentList || [];
        try {
            args.unshift(this.toStringValue());
            if (valLib[method].apply(this, args) === false) {
                this.errors.push(errorMessage || method);
            }
        }
        catch (e) {
            this.errors.push(errorMessage || e.message);
        }
        return this;
    }
    toStringValue() {
        const { value } = this;
        if (typeof value === 'string' || value instanceof Date) {
            return value;
        }
        return JSON.stringify(value);
    }
    toJSON() {
        return {
            isValid: this.isValid,
            errors: this.errors,
        };
    }
}
const OTHER_VALIDATORS = ['contains', 'equals', 'matches'];
Object.keys(valLib).forEach((name) => {
    if (name.startsWith('is') || OTHER_VALIDATORS.includes(name)) {
        // use function here to keep the correct this context
        Validator.prototype[name] = function validate(...args) {
            const error = typeof args[0] === 'string' ? args.shift() : null;
            return this.callMethod(name, error, args);
        };
    }
});
