// baqend/esm/lib/binding/Enhancer.js
const BAQEND_ID = Symbol('BaqendId');
const BAQEND_TYPE = Symbol('BaqendType');
export class Enhancer {
    /**
     * @param superClass
     * @return typeConstructor
     */
    createProxy(superClass) {
        return class Proxy extends superClass {
        };
    }
    /**
     * @param typeConstructor
     * @returns type the managed type metadata for this class
     */
    static getBaqendType(typeConstructor) {
        return typeConstructor[BAQEND_TYPE];
    }
    /**
     * @param typeConstructor
     * @return
     */
    static getIdentifier(typeConstructor) {
        return typeConstructor[BAQEND_ID];
    }
    /**
     * @param typeConstructor
     * @param identifier
     */
    static setIdentifier(typeConstructor, identifier) {
        typeConstructor[BAQEND_ID] = identifier;
    }
    /**
     * @param type
     * @param typeConstructor
     */
    enhance(type, typeConstructor) {
        if (typeConstructor[BAQEND_TYPE] === type) {
            return;
        }
        if (Object.prototype.hasOwnProperty.call(typeConstructor, BAQEND_TYPE)) {
            throw new Error('Type is already used by a different manager');
        }
        typeConstructor[BAQEND_TYPE] = type;
        Enhancer.setIdentifier(typeConstructor, type.ref);
        this.enhancePrototype(typeConstructor.prototype, type);
    }
    /**
     * Enhance the prototype of the type
     * @param proto
     * @param type
     */
    enhancePrototype(proto, type) {
        if (type.isEmbeddable) {
            return; // we do not need to enhance the prototype of embeddable
        }
        if (proto.toString === Object.prototype.toString) {
            // implements a better convenience toString method
            Object.defineProperty(proto, 'toString', {
                value: function toString() {
                    return this._metadata.id || this._metadata.bucket;
                },
                enumerable: false,
            });
        }
        // enhance all persistent object properties
        if (type.superType && type.superType.name === 'Object') {
            type.superType.declaredAttributes.forEach((attr) => {
                if (!attr.isMetadata) {
                    this.enhanceProperty(proto, attr);
                }
            });
        }
        // enhance all persistent properties
        type.declaredAttributes.forEach((attr) => {
            this.enhanceProperty(proto, attr);
        });
    }
    /**
     * @param proto
     * @param attribute
     */
    enhanceProperty(proto, attribute) {
        const { name } = attribute;
        Object.defineProperty(proto, name, {
            get() {
                this._metadata.throwUnloadedPropertyAccess(name);
                return null;
            },
            set(value) {
                this._metadata.throwUnloadedPropertyAccess(name);
                Object.defineProperty(this, name, {
                    value,
                    writable: true,
                    enumerable: true,
                    configurable: true,
                });
            },
            configurable: true,
            enumerable: true,
        });
    }
}
