// baqend/esm/lib/binding/Factory.js
// @ts-ignore
export class Factory {
    constructor() {
        this.type = null;
        this.prototype = null;
    }
    static extend(target, proto) {
        if (proto !== Factory.prototype) {
            this.extend(target, Object.getPrototypeOf(proto));
        }
        const properties = Object.getOwnPropertyNames(proto);
        for (let j = 0, len = properties.length; j < len; j += 1) {
            const prop = properties[j];
            Object.defineProperty(target, prop, Object.getOwnPropertyDescriptor(proto, prop));
        }
        return target;
    }
    /**
     * Creates a new Factory for the given type
     * @param type - the type constructor of T
     * @return A new object factory to created instances of T
     */
    static createFactory(type) {
        // We want te explicitly name the created factory and give the constructor a properly argument name
        const factory = Factory.extend((function FactoryConstructor(...args) {
            return factory.newInstance(args);
        }), this.prototype);
        // lets instanceof work properly
        factory.prototype = type.prototype;
        factory.type = type;
        return factory;
    }
    /**
     * Creates a new instance of the factory type
     * @param args Constructor arguments used for instantiation
     * @return A new created instance of *
     * @instance
     */
    new(...args) {
        return this.newInstance(args);
    }
    /**
     * Creates a new instance of the factory type
     * @param args Constructor arguments used for instantiation
     * @return A new created instance of *
     * @instance
     */
    newInstance(args) {
        if (!args || args.length === 0) {
            // eslint-disable-next-line new-cap
            return new this.type();
        }
        // es6 constructors can't be called, therefore bind all arguments and invoke the constructor
        // then with the bounded parameters
        // The first argument is shift out by invocation with `new`.
        const a = [null];
        Array.prototype.push.apply(a, args);
        const boundConstructor = (Function.prototype.bind.apply(this.type, a));
        // eslint-disable-next-line new-cap
        return new boundConstructor();
    }
}
