// baqend/esm/lib/binding/FileFactory.js
import { Factory } from './Factory';
import { File } from './File';
import * as message from '../message';
import { trailingSlashIt } from '../util';
import { StatusCode } from '../connector';
import { Permission } from '../intersection';
export class FileFactory extends Factory {
    constructor() {
        super(...arguments);
        /**
         * The owning EntityManager where this factory belongs to
         */
        this.db = null;
    }
    /**
     * Creates a new FileFactory for the given type
     * @param db
     * @return A new file factory
     */
    static create(db) {
        const factory = this.createFactory(File);
        factory.db = db;
        return factory;
    }
    /**
     * Creates a new file
     * @param args Constructor arguments used for instantiation, the constructor will not be called
     * when no arguments are passed
     * @return A new created file
     */
    newInstance(args) {
        const instance = super.newInstance(args);
        instance.db = this.db;
        return instance;
    }
    /**
     * Deserialize the file metadata from a json object back to a new file instance
     * @param json The file metadata as json
     * @return The deserialize File instance
     */
    fromJSON(json) {
        const file = this.newInstance([json.id]);
        file.fromJSON(json);
        return file;
    }
    /**
     * Updates the metadata of the root file directory formally the file "bucket"
     * @param bucket The name of the root file directory
     * @param metadata The new metadata for the bucket
     * @param doneCallback Invoked if the operation succeeds
     * @param failCallback The callback is invoked if any error has occurred
     * @return A promise which will fulfilled with the updated metadata
     */
    saveMetadata(bucket, metadata, doneCallback, failCallback) {
        const msg = new message.SetFileBucketMetadata(bucket, metadata);
        return this.db.send(msg).then(doneCallback, failCallback);
    }
    /**
     * Gets the metadata of the root folder (formally the file "bucket")
     * @param bucket The name of the root file directory
     * @param options The load metadata options
     * @param [options.refresh=false] Force a revalidation while fetching the metadata
     * @param doneCallback
     * The callback is invoked after the metadata is fetched
     * @param failCallback The callback is invoked if any error has occurred
     * @return A promise which will be fulfilled with the bucket ACLs
     */
    loadMetadata(bucket, options, doneCallback, failCallback) {
        const msg = new message.GetFileBucketMetadata(bucket);
        // this._db.ensureCacheHeader(this.id, msg, options.refresh);
        // do not white list the file, because head-request does not revalidate the cache.
        return this.db.send(msg).then((response) => {
            const result = {};
            Permission.BASE_PERMISSIONS.forEach((key) => {
                result[key] = Permission.fromJSON(response.entity[key] || {});
            });
            return result;
        }, (e) => {
            if (e.status === StatusCode.OBJECT_NOT_FOUND) {
                return null;
            }
            throw e;
        }).then(doneCallback, failCallback);
    }
    /**
     * Lists all the buckets.
     * @param doneCallback The callback is invoked with the listed buckets
     * @param failCallback The callback is invoked if any error has occurred
     * @return The listed buckets.
     */
    listBuckets(doneCallback, failCallback) {
        return this.db.send(new message.ListBuckets()).then((response) => (response.entity.map((bucket) => this.new(`${bucket}/`)))).then(doneCallback, failCallback);
    }
    /**
     * Lists the files (and folders) in the given folder.
     *
     * @param folderOrPath The folder/path to list.
     * @param start The file/folder from where to start listing (not included)
     * @param count The maximum number of files to return.
     * @param doneCallback The callback is invoked with the listed files
     * @param failCallback The callback is invoked if any error has occurred
     * @return The listed files/folders.
     */
    listFiles(folderOrPath, start, count, doneCallback, failCallback) {
        let folder;
        if (typeof folderOrPath === 'string') {
            const path = trailingSlashIt(folderOrPath);
            folder = this.new({ path });
        }
        else {
            folder = folderOrPath;
        }
        const path = folder.key;
        const { bucket } = folder;
        return this.db.send(new message.ListFiles(bucket, path, start ? start.key : undefined, count)).then((response) => (response.entity.map((file) => this.new(file)))).then(doneCallback, failCallback);
    }
}
