// baqend/esm/lib/binding/ManagedFactory.js
import { Factory } from './Factory';
import { Metadata } from '../intersection';
export class ManagedFactory extends Factory {
    constructor() {
        super(...arguments);
        /**
         * Methods that are added to object instances
         * This property is an alias for this factory type prototype
         * @name methods
         */
        this.methods = null;
        /**
         * The managed type of this factory
         */
        this.managedType = null;
        /**
         * The owning EntityManager where this factory belongs to
         */
        this.db = null;
    }
    /**
     * Creates a new ManagedFactory for the given type
     * @param managedType The metadata of type T
     * @param db The entity manager instance
     */
    static create(managedType, db) {
        const factory = this.createFactory(managedType.typeConstructor);
        factory.methods = factory.prototype;
        factory.managedType = managedType;
        factory.db = db;
        return factory;
    }
    /**
     * Creates a new instance and sets the Managed Object to the given json
     * @param json
     * @return A new created instance of T
     */
    fromJSON(json) {
        const instance = this.newInstance();
        return this.managedType.fromJsonValue(Metadata.create(this.managedType, this.db), json, instance, {
            persisting: false,
        });
    }
    /**
     * Adds methods to instances of this factories type
     * @param methods The methods to add
     * @return
     */
    addMethods(methods) {
        Object.assign(this.methods, methods);
    }
    /**
     * Add a method to instances of this factories type
     * @param name The method name to add
     * @param fn The Method to add
     * @return
     */
    addMethod(name, fn) {
        this.methods[name] = fn;
    }
}
