// baqend/commonjs/lib/metamodel/ModelBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelBuilder = void 0;
var BasicType_1 = require("./BasicType");
var EntityType_1 = require("./EntityType");
var EmbeddableType_1 = require("./EmbeddableType");
var ListAttribute_1 = require("./ListAttribute");
var MapAttribute_1 = require("./MapAttribute");
var SetAttribute_1 = require("./SetAttribute");
var SingularAttribute_1 = require("./SingularAttribute");
var error_1 = require("../error");
var intersection_1 = require("../intersection");
var ModelBuilder = /** @class */ (function () {
    function ModelBuilder() {
        var _this = this;
        this.models = {};
        this.modelDescriptors = null;
        Object.keys(BasicType_1.BasicType).forEach(function (typeName) {
            var basicType = BasicType_1.BasicType[typeName];
            if (basicType instanceof BasicType_1.BasicType) {
                _this.models[basicType.ref] = basicType;
            }
        });
    }
    /**
     * @param ref
     * @return
     */
    ModelBuilder.prototype.getModel = function (ref) {
        if (ref in this.models) {
            return this.models[ref];
        }
        var model = this.buildModel(ref);
        this.models[ref] = model;
        return model;
    };
    /**
     * @param modelDescriptors
     * @return
     */
    ModelBuilder.prototype.buildModels = function (modelDescriptors) {
        var _this = this;
        this.modelDescriptors = {};
        modelDescriptors.forEach(function (modelDescriptor) {
            _this.modelDescriptors[modelDescriptor.class] = modelDescriptor;
        });
        Object.keys(this.modelDescriptors).forEach(function (ref) {
            try {
                var model = _this.getModel(ref);
                _this.buildAttributes(model);
            }
            catch (e) {
                throw new error_1.PersistentError("Can't create model for entity class " + ref, e);
            }
        });
        // ensure at least an object entity
        this.getModel(EntityType_1.EntityType.Object.ref);
        return this.models;
    };
    /**
     * @param ref
     * @return
     */
    ModelBuilder.prototype.buildModel = function (ref) {
        var modelDescriptor = this.modelDescriptors[ref];
        var type;
        if (ref === EntityType_1.EntityType.Object.ref) {
            type = new EntityType_1.EntityType.Object();
        }
        else if (modelDescriptor) {
            if (modelDescriptor.embedded) {
                type = new EmbeddableType_1.EmbeddableType(ref);
            }
            else {
                var superTypeIdentifier = modelDescriptor.superClass || EntityType_1.EntityType.Object.ref;
                type = new EntityType_1.EntityType(ref, this.getModel(superTypeIdentifier));
            }
        }
        else {
            throw new TypeError("No model available for " + ref);
        }
        type.metadata = {};
        if (modelDescriptor) {
            type.metadata = modelDescriptor.metadata || {};
            var permissions_1 = modelDescriptor.acl || {};
            Object.keys(permissions_1).forEach(function (permission) {
                var permissionProperty = permission + "Permission";
                type[permissionProperty].fromJSON(permissions_1[permission]);
            });
        }
        return type;
    };
    /**
     * @param model
     * @return
     */
    ModelBuilder.prototype.buildAttributes = function (model) {
        var _this = this;
        var modelDescriptor = this.modelDescriptors[model.ref];
        var fields = modelDescriptor.fields;
        Object.keys(fields).forEach(function (name) {
            var field = fields[name];
            if (!model.getAttribute(name)) { // skip predefined attributes
                model.addAttribute(_this.buildAttribute(field), field.order);
            }
        });
        if (typeof modelDescriptor.validationCode === 'string') {
            // eslint-disable-next-line no-param-reassign
            model.validationCode = intersection_1.Validator.compile(model, modelDescriptor.validationCode);
        }
    };
    /**
     * @param field The field metadata
     * @param field.name The name of zhe field
     * @param field.type The type reference of the field
     * @param field.order The order number of the field
     * @param field.metadata Additional metadata of the field
     * @return
     */
    ModelBuilder.prototype.buildAttribute = function (field) {
        // TODO: remove readonly if createdAt and updatedAt becomes real metadata fields in the schema
        var isMetadata = field.flags && (field.flags.indexOf('METADATA') !== -1 || field.flags.indexOf('READONLY') !== -1);
        var name = field.name;
        var ref = field.type;
        if (ref.indexOf('/db/collection.') !== 0) {
            var singularAttribute = new SingularAttribute_1.SingularAttribute(name, this.getModel(ref), isMetadata);
            singularAttribute.metadata = field.metadata;
            return singularAttribute;
        }
        var collectionType = ref.substring(0, ref.indexOf('['));
        var elementType = ref.substring(ref.indexOf('[') + 1, ref.indexOf(']')).trim();
        switch (collectionType) {
            case ListAttribute_1.ListAttribute.ref:
                return new ListAttribute_1.ListAttribute(name, this.getModel(elementType));
            case SetAttribute_1.SetAttribute.ref:
                return new SetAttribute_1.SetAttribute(name, this.getModel(elementType));
            case MapAttribute_1.MapAttribute.ref: {
                var keyType = elementType.substring(0, elementType.indexOf(',')).trim();
                var valueType = elementType.substring(elementType.indexOf(',') + 1).trim();
                return new MapAttribute_1.MapAttribute(name, this.getModel(keyType), this.getModel(valueType));
            }
            default:
                throw new TypeError("No collection available for " + ref);
        }
    };
    return ModelBuilder;
}());
exports.ModelBuilder = ModelBuilder;
