// baqend/commonjs/lib/binding/Enhancer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enhancer = void 0;
var BAQEND_ID = Symbol('BaqendId');
var BAQEND_TYPE = Symbol('BaqendType');
var Enhancer = /** @class */ (function () {
    function Enhancer() {
    }
    /**
     * @param superClass
     * @return typeConstructor
     */
    Enhancer.prototype.createProxy = function (superClass) {
        return /** @class */ (function (_super) {
            __extends(Proxy, _super);
            function Proxy() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            return Proxy;
        }(superClass));
    };
    /**
     * @param typeConstructor
     * @returns type the managed type metadata for this class
     */
    Enhancer.getBaqendType = function (typeConstructor) {
        return typeConstructor[BAQEND_TYPE];
    };
    /**
     * @param typeConstructor
     * @return
     */
    Enhancer.getIdentifier = function (typeConstructor) {
        return typeConstructor[BAQEND_ID];
    };
    /**
     * @param typeConstructor
     * @param identifier
     */
    Enhancer.setIdentifier = function (typeConstructor, identifier) {
        typeConstructor[BAQEND_ID] = identifier;
    };
    /**
     * @param type
     * @param typeConstructor
     */
    Enhancer.prototype.enhance = function (type, typeConstructor) {
        if (typeConstructor[BAQEND_TYPE] === type) {
            return;
        }
        if (Object.prototype.hasOwnProperty.call(typeConstructor, BAQEND_TYPE)) {
            throw new Error('Type is already used by a different manager');
        }
        typeConstructor[BAQEND_TYPE] = type;
        Enhancer.setIdentifier(typeConstructor, type.ref);
        this.enhancePrototype(typeConstructor.prototype, type);
    };
    /**
     * Enhance the prototype of the type
     * @param proto
     * @param type
     */
    Enhancer.prototype.enhancePrototype = function (proto, type) {
        var _this = this;
        if (type.isEmbeddable) {
            return; // we do not need to enhance the prototype of embeddable
        }
        if (proto.toString === Object.prototype.toString) {
            // implements a better convenience toString method
            Object.defineProperty(proto, 'toString', {
                value: function toString() {
                    return this._metadata.id || this._metadata.bucket;
                },
                enumerable: false,
            });
        }
        // enhance all persistent object properties
        if (type.superType && type.superType.name === 'Object') {
            type.superType.declaredAttributes.forEach(function (attr) {
                if (!attr.isMetadata) {
                    _this.enhanceProperty(proto, attr);
                }
            });
        }
        // enhance all persistent properties
        type.declaredAttributes.forEach(function (attr) {
            _this.enhanceProperty(proto, attr);
        });
    };
    /**
     * @param proto
     * @param attribute
     */
    Enhancer.prototype.enhanceProperty = function (proto, attribute) {
        var name = attribute.name;
        Object.defineProperty(proto, name, {
            get: function () {
                this._metadata.throwUnloadedPropertyAccess(name);
                return null;
            },
            set: function (value) {
                this._metadata.throwUnloadedPropertyAccess(name);
                Object.defineProperty(this, name, {
                    value: value,
                    writable: true,
                    enumerable: true,
                    configurable: true,
                });
            },
            configurable: true,
            enumerable: true,
        });
    };
    return Enhancer;
}());
exports.Enhancer = Enhancer;
